<!doctype html>
<html lang="fr">
<head>
<title>Séquençage WS et SW</title>
<!-- 2016-03-30 Wed 17:38 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta name="generator" content="Org-mode">
<meta name="author" content="Samuel Barreto">

<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.6/paper/bootstrap.min.css" rel="stylesheet" integrity="sha384-2mX2PSpkRSXLQzmNzH3gwK6srb06+OfbDlYjbog8LQuALYJjuQ3+Yzy2JIWNV9rW" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/1.11.3/jquery.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.5/js/bootstrap.min.js"></script>
<style type="text/css">
/* org mode styles on top of twbs */

html {
    position: relative;
    min-height: 100%;
}

body {
    font-size: 18px;
    margin-bottom: 105px;
}

footer {
    position: absolute;
    bottom: 0;
    width: 100%;
    height: 101px;
    background-color: #f5f5f5;
}

footer > div {
    padding: 10px;
}

footer p {
    margin: 0 0 5px;
    text-align: center;
    font-size: 16px;
}

#table-of-contents {
    margin-top: 20px;
    margin-bottom: 20px;
}

blockquote p {
    font-size: 18px;
}

pre {
    font-size: 16px;
}

.footpara {
    display: inline-block;
}

figcaption {
  font-size: 16px;
  color: #666;
  font-style: italic;
  padding-bottom: 15px;
}

/* from twbs docs */

.bs-docs-sidebar.affix {
    position: static;
}
@media (min-width: 768px) {
    .bs-docs-sidebar {
        padding-left: 20px;
    }
}

/* All levels of nav */
.bs-docs-sidebar .nav > li > a {
    display: block;
    padding: 4px 20px;
    font-size: 14px;
    font-weight: 500;
    color: #999;
}
.bs-docs-sidebar .nav > li > a:hover,
.bs-docs-sidebar .nav > li > a:focus {
    padding-left: 19px;
    color: #A1283B;
    text-decoration: none;
    background-color: transparent;
    border-left: 1px solid #A1283B;
}
.bs-docs-sidebar .nav > .active > a,
.bs-docs-sidebar .nav > .active:hover > a,
.bs-docs-sidebar .nav > .active:focus > a {
    padding-left: 18px;
    font-weight: bold;
    color: #A1283B;
    background-color: transparent;
    border-left: 2px solid #A1283B;
}

/* Nav: second level (shown on .active) */
.bs-docs-sidebar .nav .nav {
    display: none; /* Hide by default, but at >768px, show it */
    padding-bottom: 10px;
}
.bs-docs-sidebar .nav .nav > li > a {
    padding-top: 1px;
    padding-bottom: 1px;
    padding-left: 30px;
    font-size: 12px;
    font-weight: normal;
}
.bs-docs-sidebar .nav .nav > li > a:hover,
.bs-docs-sidebar .nav .nav > li > a:focus {
    padding-left: 29px;
}
.bs-docs-sidebar .nav .nav > .active > a,
.bs-docs-sidebar .nav .nav > .active:hover > a,
.bs-docs-sidebar .nav .nav > .active:focus > a {
    padding-left: 28px;
    font-weight: 500;
}

/* Nav: third level (shown on .active) */
.bs-docs-sidebar .nav .nav .nav {
    padding-bottom: 10px;
}
.bs-docs-sidebar .nav .nav .nav > li > a {
    padding-top: 1px;
    padding-bottom: 1px;
    padding-left: 40px;
    font-size: 12px;
    font-weight: normal;
}
.bs-docs-sidebar .nav .nav .nav > li > a:hover,
.bs-docs-sidebar .nav .nav .nav > li > a:focus {
    padding-left: 39px;
}
.bs-docs-sidebar .nav .nav .nav > .active > a,
.bs-docs-sidebar .nav .nav .nav > .active:hover > a,
.bs-docs-sidebar .nav .nav .nav > .active:focus > a {
    padding-left: 38px;
    font-weight: 500;
}

/* Show and affix the side nav when space allows it */
@media (min-width: 992px) {
    .bs-docs-sidebar .nav > .active > ul {
        display: block;
    }
    /* Widen the fixed sidebar */
    .bs-docs-sidebar.affix,
    .bs-docs-sidebar.affix-bottom {
        width: 213px;
    }
    .bs-docs-sidebar.affix {
        position: fixed; /* Undo the static from mobile first approach */
        top: 20px;
    }
    .bs-docs-sidebar.affix-bottom {
        position: absolute; /* Undo the static from mobile first approach */
    }
    .bs-docs-sidebar.affix .bs-docs-sidenav,.bs-docs-sidebar.affix-bottom .bs-docs-sidenav {
        margin-top: 0;
        margin-bottom: 0
    }
}
@media (min-width: 1200px) {
    /* Widen the fixed sidebar again */
    .bs-docs-sidebar.affix-bottom,
    .bs-docs-sidebar.affix {
        width: 263px;
    }
}
</style>
<script type="text/javascript">
$(function() {
    'use strict';

    $('.bs-docs-sidebar li').first().addClass('active');

    $(document.body).scrollspy({target: '.bs-docs-sidebar'});

    $('.bs-docs-sidebar').affix();
});
</script>
</head>
<body>
<div id="content" class="container">
<div class="row"><div class="col-md-9"><h1 class="title">Séquençage WS et SW</h1>
<p>
Ce dossier contient les séquences des plaques obtenues avec les
constructions SW et WS, envoyées à séquencer le <span class="timestamp-wrapper"><span class="timestamp">[2016-03-15 Tue]</span></span>.
</p>

<div id="outline-container-sec-1" class="outline-2">
<h2 id="sec-1"><span class="section-number-2">1</span> Dossiers et fichiers</h2>
<div class="outline-text-2" id="text-1">
<p>
Les données brutes au format .zip sont dans le dossier <code>raw</code>. Elles
sont extraites dans le dossier <code>data</code>, séparé en <code>ws</code> et <code>sw</code> pour les
analyses. Rien ne doit toucher au dossier <code>raw</code>.
</p>

<p>
J'ai ensuite utilisé le script <a href="./src/sort_into_dir.sh">./src/sort_into_dir.sh</a> pour classer les
fichiers selon leur extension.
</p>

<div class="org-src-container">

<pre class="src src-sh"><span style="color: #93a1a1;">#</span><span style="color: #93a1a1;">!/bin/</span><span style="color: #859900; font-weight: bold;">bash</span>

<span style="color: #657b83; font-weight: bold;">cd</span> data/ws

mkdir fasta spectro seq

find . -name <span style="color: #2aa198;">"*.fas"</span>  -maxdepth 1 -exec mv <span style="color: #2aa198;">{}</span> ./fasta/ <span style="color: #2aa198;">\;</span>
find . -name <span style="color: #2aa198;">"*.ab1"</span>  -maxdepth 1 -exec mv <span style="color: #2aa198;">{}</span> ./spectro/ <span style="color: #2aa198;">\;</span>
find . -name <span style="color: #2aa198;">"*.seq"</span>  -maxdepth 1 -exec mv <span style="color: #2aa198;">{}</span> ./seq/ <span style="color: #2aa198;">\;</span>
find . -name <span style="color: #2aa198;">"*.csv"</span>  -maxdepth 1 -exec mv <span style="color: #2aa198;">{}</span> ../csv/ <span style="color: #2aa198;">\;</span>

<span style="color: #657b83; font-weight: bold;">cd</span> ../sw

mkdir fasta spectro seq

find . -name <span style="color: #2aa198;">"*.fas"</span>  -maxdepth 1 -exec mv <span style="color: #2aa198;">{}</span> ./fasta/ <span style="color: #2aa198;">\;</span>
find . -name <span style="color: #2aa198;">"*.ab1"</span>  -maxdepth 1 -exec mv <span style="color: #2aa198;">{}</span> ./spectro/ <span style="color: #2aa198;">\;</span>
find . -name <span style="color: #2aa198;">"*.seq"</span>  -maxdepth 1 -exec mv <span style="color: #2aa198;">{}</span> ./seq/ <span style="color: #2aa198;">\;</span>
find . -name <span style="color: #2aa198;">"*.csv"</span>  -maxdepth 1 -exec mv <span style="color: #2aa198;">{}</span> ../csv/ <span style="color: #2aa198;">\;</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-sec-2" class="outline-2">
<h2 id="sec-2"><span class="section-number-2">2</span> Contrôle qualité</h2>
<div class="outline-text-2" id="text-2">
<p>
Après avoir reçu les séquences, j'ai voulu connaître la qualité
globale des séquences. Je les ai donc converties en fastq avec le
script <a href="./src/ab1_to_fastq.sh">./src/ab1_to_fastq.sh</a>, et j'ai utilisé <code>fastqc</code> pour
représenter la qualité globale via le script <a href="./src/quality_control.sh">./src/quality_control.sh</a>.
</p>

<div class="org-src-container">

<pre class="src src-sh"><span style="color: #93a1a1;">#</span><span style="color: #93a1a1;">!/bin/</span><span style="color: #859900; font-weight: bold;">bash</span>

<span style="color: #93a1a1;"># </span><span style="color: #93a1a1;">transforme en fastq</span>
<span style="color: #657b83; font-weight: bold;">cd</span> data/ws

touch ws_untrimmed.fastq

<span style="color: #859900; font-weight: bold;">for</span> file<span style="color: #859900; font-weight: bold;"> in</span> spectro/*.ab1
<span style="color: #859900; font-weight: bold;">do</span>
    seqret <span style="color: #b58900; font-weight: bold;">\</span>
        -sformat abi <span style="color: #b58900; font-weight: bold;">\</span>
        -osformat fastq <span style="color: #b58900; font-weight: bold;">\</span>
        -auto <span style="color: #b58900; font-weight: bold;">\</span>
        -stdout <span style="color: #b58900; font-weight: bold;">\</span>
        -sequence $<span style="color: #268bd2;">file</span> <span style="color: #b58900; font-weight: bold;">\</span>
        &gt;&gt; ws_untrimmed.fastq
<span style="color: #859900; font-weight: bold;">done</span>

<span style="color: #657b83; font-weight: bold;">cd</span> ../sw

<span style="color: #93a1a1;"># </span><span style="color: #93a1a1;">TEMPORAIRE, &#233;limine le fichier .ab1 tendancieux pour l'instant</span>
<span style="color: #93a1a1;"># </span><span style="color: #93a1a1;">rm spectro/psw76-1073bis.ab1</span>

touch sw_untrimmed.fastq

<span style="color: #859900; font-weight: bold;">for</span> file<span style="color: #859900; font-weight: bold;"> in</span> spectro/*.ab1
<span style="color: #859900; font-weight: bold;">do</span>
    seqret <span style="color: #b58900; font-weight: bold;">\</span>
        -sformat abi <span style="color: #b58900; font-weight: bold;">\</span>
        -osformat fastq <span style="color: #b58900; font-weight: bold;">\</span>
        -auto <span style="color: #b58900; font-weight: bold;">\</span>
        -stdout <span style="color: #b58900; font-weight: bold;">\</span>
        -sequence $<span style="color: #268bd2;">file</span> <span style="color: #b58900; font-weight: bold;">\</span>
        &gt;&gt; sw_untrimmed.fastq
<span style="color: #859900; font-weight: bold;">done</span>
</pre>
</div>

<p>
J'ai utilisé <code>fastqc</code> pour représenter la qualité globale via le
script <a href="./src/quality_control.sh">./src/quality_control.sh</a>.
</p>

<div class="org-src-container">

<pre class="src src-sh"><span style="color: #93a1a1;">#</span><span style="color: #93a1a1;">!/bin/</span><span style="color: #859900; font-weight: bold;">bash</span>

<span style="color: #93a1a1;"># </span><span style="color: #93a1a1;">contr&#244;le qualit&#233; des s&#233;quences.</span>
<span style="color: #657b83; font-weight: bold;">cd</span> data

<span style="color: #93a1a1;"># </span><span style="color: #93a1a1;">cr&#233;e un dossier qc de quality control pour ws et sw</span>
mkdir -p qc/<span style="color: #2aa198;">{</span>ws,sw<span style="color: #2aa198;">}</span>

<span style="color: #93a1a1;"># </span><span style="color: #93a1a1;">fastqc sortie dasn le dossier appropri&#233;</span>
fastqc ws/ws_untrimmed.fastq -o qc/ws
fastqc sw/sw_untrimmed.fastq -o qc/sw

<span style="color: #93a1a1;"># </span><span style="color: #93a1a1;">transfert du fichier .png qui nous int&#233;resse dans le dir adapt&#233;</span>
<span style="color: #93a1a1;"># </span><span style="color: #93a1a1;">et suppression des dossiers inutiles</span>
<span style="color: #93a1a1;">#</span><span style="color: #93a1a1;">' -qq : very quiet</span>
<span style="color: #93a1a1;">#</span><span style="color: #93a1a1;">' -o  : overwrite without warning</span>
<span style="color: #93a1a1;">#</span><span style="color: #93a1a1;">' -d  : out dir</span>
unzip -qq -o <span style="color: #b58900; font-weight: bold;">\</span>
      qc/sw/sw_untrimmed_fastqc.zip <span style="color: #b58900; font-weight: bold;">\</span>
      sw_untrimmed_fastqc/Images/per_base_quality.png <span style="color: #b58900; font-weight: bold;">\</span>
      -d ../anl/ <span style="color: #b58900; font-weight: bold;">\</span>
    &amp;&amp; mv ../anl/sw_untrimmed_fastqc/Images/per_base_quality.png <span style="color: #b58900; font-weight: bold;">\</span>
          ../anl/sw_per_base_qual.png <span style="color: #b58900; font-weight: bold;">\</span>
    &amp;&amp; rm -r ../anl/sw_untrimmed_fastqc

<span style="color: #93a1a1;"># </span><span style="color: #93a1a1;">transfert du fichier .png qui nous int&#233;resse dans le dir adapt&#233;</span>
<span style="color: #93a1a1;"># </span><span style="color: #93a1a1;">et suppression des dossiers inutiles</span>
unzip -qq -o <span style="color: #b58900; font-weight: bold;">\</span>
      qc/ws/ws_untrimmed_fastqc.zip <span style="color: #b58900; font-weight: bold;">\</span>
      ws_untrimmed_fastqc/Images/per_base_quality.png <span style="color: #b58900; font-weight: bold;">\</span>
      -d ../anl/ <span style="color: #b58900; font-weight: bold;">\</span>
    &amp;&amp; mv ../anl/ws_untrimmed_fastqc/Images/per_base_quality.png <span style="color: #b58900; font-weight: bold;">\</span>
          ../anl/ws_per_base_qual.png <span style="color: #b58900; font-weight: bold;">\</span>
    &amp;&amp; rm -r ../anl/ws_untrimmed_fastqc
</pre>
</div>

<p>
À priori les séquences sont plutôt de bonne qualité. Voir le résumé
dans le fichier <a href="./data/qc/sw/sw_untrimmed_fastqc.html">./data/qc/sw/sw_untrimmed_fastqc.html</a> et
<a href="./data/qc/ws/ws_untrimmed_fastqc.html">./data/qc/ws/ws_untrimmed_fastqc.html</a>.
</p>
</div>
</div>

<div id="outline-container-sec-3" class="outline-2">
<h2 id="sec-3"><span class="section-number-2">3</span> Analyse rapide du SNP calling de GATC</h2>
<div class="outline-text-2" id="text-3">
<p>
J'ai rapidement analysé la sortie du SNP calling effectué par GATC,
mais il manque des données. Je vais refaire le SNP calling moi-même,
en utilisant dans un premier temps ssaha2.
</p>

<p>
Le script est là <a href="./anl/gatc_snpcall/snp_call.r">./anl/gatc_snpcall/snp_call.r</a>, les sorties en
html là <a href="./anl/gatc_snpcall/snp_call.html">./anl/gatc_snpcall/snp_call.html</a>.
</p>
</div>
</div>

<div id="outline-container-sec-4" class="outline-2">
<h2 id="sec-4"><span class="section-number-2">4</span> SNP calling global</h2>
<div class="outline-text-2" id="text-4">
</div><div id="outline-container-sec-4-1" class="outline-3">
<h3 id="sec-4-1"><span class="section-number-3">4.1</span> trimming des séquences</h3>
<div class="outline-text-3" id="text-4-1">
<p>
J'ai utilisé le programme bbduk pour trimmer les séquences de faible
qualité.
</p>

<div class="org-src-container">

<pre class="src src-sh">  <span style="color: #93a1a1;">#</span><span style="color: #93a1a1;">!/bin/bash</span>

  <span style="color: #93a1a1;">#</span><span style="color: #93a1a1;">' -qtrim=rl : quality trim right and left</span>
  <span style="color: #93a1a1;">#</span><span style="color: #93a1a1;">' -trimq=28 : trim if quality &lt; 28 (sanger encoding, illumina 1.9)</span>
  <span style="color: #93a1a1;">#</span><span style="color: #93a1a1;">' -minlen=620 : keep only seq with length &gt; 620, after trimming.</span>
  <span style="color: #93a1a1;">#</span><span style="color: #93a1a1;">' -Xmx1g : tells bbduk / java to use 1G of RAM</span>

  <span style="color: #268bd2;">FASTQ</span>=<span style="color: #2aa198;">"data/sw/sw_untrimmed.fastq</span>
<span style="color: #2aa198;">  data/ws/ws_untrimmed.fastq</span>
<span style="color: #2aa198;">  "</span>

  <span style="color: #859900; font-weight: bold;">for</span> f<span style="color: #859900; font-weight: bold;"> in</span> $<span style="color: #268bd2;">FASTQ</span>
  <span style="color: #859900; font-weight: bold;">do</span>
      <span style="color: #268bd2;">out</span>=$<span style="color: #268bd2;">f</span>.trim
      bin/bbmap/bbduk.sh -Xmx1g <span style="color: #b58900; font-weight: bold;">\</span>
                         -in=$<span style="color: #268bd2;">f</span> <span style="color: #b58900; font-weight: bold;">\</span>
                         -out=$<span style="color: #268bd2;">f</span>.qtrim <span style="color: #b58900; font-weight: bold;">\</span>
                         -qtrim=rl <span style="color: #b58900; font-weight: bold;">\</span>
                         -trimq=28 <span style="color: #b58900; font-weight: bold;">\</span>
                         -minlen=600

      <span style="color: #93a1a1;">## </span><span style="color: #93a1a1;">convertit les bases d'une qualit&#233; inf&#233;rieure &#224; 20 en N.</span>
      seqtk seq <span style="color: #b58900; font-weight: bold;">\</span>
            -q20 <span style="color: #b58900; font-weight: bold;">\</span>
            -nN <span style="color: #b58900; font-weight: bold;">\</span>
            $<span style="color: #268bd2;">f</span>.qtrim &gt; $<span style="color: #268bd2;">out</span>

      <span style="color: #93a1a1;">## </span><span style="color: #93a1a1;">convertit le fastq en fasta</span>
      seqret <span style="color: #b58900; font-weight: bold;">\</span>
          -sformat fastq <span style="color: #b58900; font-weight: bold;">\</span>
          -osformat fasta <span style="color: #b58900; font-weight: bold;">\</span>
          -auto -stdout <span style="color: #b58900; font-weight: bold;">\</span>
          -sequence $<span style="color: #268bd2;">out</span> &gt; $<span style="color: #268bd2;">f</span>.fasta

      rm $<span style="color: #268bd2;">f</span>.qtrim
  <span style="color: #859900; font-weight: bold;">done</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-sec-4-2" class="outline-3">
<h3 id="sec-4-2"><span class="section-number-3">4.2</span> Contrôle qualité des séquences trimmées</h3>
<div class="outline-text-3" id="text-4-2">
<div class="org-src-container">

<pre class="src src-sh">  <span style="color: #93a1a1;">#</span><span style="color: #93a1a1;">!/usr/bin/env bash</span>

  <span style="color: #657b83; font-weight: bold;">cd</span> data

  <span style="color: #268bd2;">TRIM</span>=<span style="color: #2aa198;">"ws/ws_untrimmed.fastq.trim sw/sw_untrimmed.fastq.trim"</span>

  mkdir qtrim_qc

  <span style="color: #859900; font-weight: bold;">for</span> f<span style="color: #859900; font-weight: bold;"> in</span> $<span style="color: #268bd2;">TRIM</span>
  <span style="color: #859900; font-weight: bold;">do</span>
      fastqc $<span style="color: #268bd2;">f</span> -o qtrim_qc
  <span style="color: #859900; font-weight: bold;">done</span>
</pre>
</div>

<p>
Le script est là <a href="./src/qtrim_qc.sh">./src/qtrim_qc.sh</a> et les sorties sont là
<a href="data/qtrim_qc/sw_untrimmed.fastq.trim_fastqc.html">./data/qtrim<sub>qc</sub>/sw<sub>untrimmed.fastq.trim</sub><sub>fastqc.html</sub></a> et là
<a href="data/qtrim_qc/sw_untrimmed.fastq.trim_fastqc.html">./data/qtrim<sub>qc</sub>/ws<sub>untrimmed.fastq.trim</sub><sub>fastqc.html</sub></a>.
</p>

<p>
Les séquences sont nettement raccourcies par rapport aux séquences non trimmées. Mais l'ensemble des
bases qu'on veut analyser sont de bonne qualité. C'est donc plutôt bon signe.
</p>
</div>
</div>

<div id="outline-container-sec-4-3" class="outline-3">
<h3 id="sec-4-3"><span class="section-number-3">4.3</span> Pooling des séquences dans un seul fastq</h3>
<div class="outline-text-3" id="text-4-3">
<div class="org-src-container">

<pre class="src src-sh">  <span style="color: #93a1a1;">#</span><span style="color: #93a1a1;">!/usr/bin/env bash</span>

  cat data/sw/sw_untrimmed.fastq.trim data/ws/ws_untrimmed.fastq.trim &gt; data/trimmed.fastq
</pre>
</div>

<p>
<a href="./src/pool_trim.sh">./src/pool_trim.sh</a>
</p>
</div>
</div>

<div id="outline-container-sec-4-4" class="outline-3">
<h3 id="sec-4-4"><span class="section-number-3">4.4</span> Analyses des alignements via seaview</h3>
<div class="outline-text-3" id="text-4-4">
<p>
Pour les analyses dans seaview, il faut un fichier fasta. Je lui
ajoute la référence.
</p>

<div class="org-src-container">

<pre class="src src-sh"><span style="color: #93a1a1;">#</span><span style="color: #93a1a1;">!/usr/bin/</span><span style="color: #859900; font-weight: bold;">env</span><span style="color: #93a1a1;"> bash</span>

seqret -sformat fastq -osformat fasta -auto -stdout <span style="color: #b58900; font-weight: bold;">\</span>
       -sequence data/trimmed.fastq &gt; data/trimmed.fasta

fastx_reverse_complement <span style="color: #b58900; font-weight: bold;">\</span>
    -i raw/ref/reference1073bis-1392.fasta <span style="color: #b58900; font-weight: bold;">\</span>
    &gt;&gt; data/trimmed.fasta
</pre>
</div>

<p>
<a href="./src/trim_to_fasta.sh">./src/trim_to_fasta.sh</a>
</p>

<p>
J'ai placé les différentes analyses graphiques via seaview dans le
dossier <a href="./anl/align">./anl/align</a>.
</p>

<p>
Différents alignements :
</p>
<ul class="org-ul">
<li>l'alignement global : <a href="./anl/align/trimmed.mase">./anl/align/trimmed.mase</a>
</li>
<li>l'alignement avec seulement les sites variables : <a href="./anl/align/variable_site.mase">./anl/align/variable_site.mase</a>
</li>
</ul>

<p>
Sur l'alignement global, j'ai repéré quelques cas de traces complexes,
où des séquences avaient ‘gardé’ leur allèle sauvage, quand les deux
SNP environnants sont les SNP transformants.
</p>
</div>
</div>

<div id="outline-container-sec-4-5" class="outline-3">
<h3 id="sec-4-5"><span class="section-number-3">4.5</span> snp calling via ssahaSNP</h3>
<div class="outline-text-3" id="text-4-5">
<p>
Pour installer ssahaSNP voir les instructions là :
<a href="ftp://ftp.sanger.ac.uk/pub/resources/software/ssahasnp/">ftp://ftp.sanger.ac.uk/pub/resources/software/ssahasnp/</a>.
</p>

<p>
Il est très important que les fichiers fasta de référence soit formattés de la
bonne façon, sinon ssahaSNP ne sort pas les bonnes valeurs dans les lignes de
tableau.
</p>

<div class="org-src-container">

<pre class="src src-sh">   <span style="color: #93a1a1;">#</span><span style="color: #93a1a1;">!/usr/bin/env bash</span>

   <span style="color: #93a1a1;"># </span><span style="color: #93a1a1;">Variant calling using ssaha2 and ssahaSNP</span>

   <span style="color: #859900; font-weight: bold;">if</span> <span style="color: #2aa198;">[</span> <span style="color: #b58900; font-weight: bold;">!</span> -e data/snp_call <span style="color: #2aa198;">]</span>
   <span style="color: #859900; font-weight: bold;">then</span>
       mkdir data/snp_call
   <span style="color: #859900; font-weight: bold;">fi</span>

   <span style="color: #657b83; font-weight: bold;">cd</span> data/snp_call

   <span style="color: #859900; font-weight: bold;">if</span> <span style="color: #2aa198;">[</span><span style="color: #b58900;">[</span> <span style="color: #b58900; font-weight: bold;">!</span> -e trimmed.fastq &amp;&amp; <span style="color: #b58900; font-weight: bold;">!</span> -e ref.fasta <span style="color: #b58900;">]</span><span style="color: #2aa198;">]</span>
   <span style="color: #859900; font-weight: bold;">then</span>
       ln -s    ../trimmed.fastq .
       ln -s ../../raw/ref/reference1073bis-1392.fasta ref.fasta
   <span style="color: #859900; font-weight: bold;">fi</span>

   <span style="color: #93a1a1;">## </span><span style="color: #93a1a1;">alignement &#224; la s&#233;quence de r&#233;f&#233;rence</span>
   <span style="color: #93a1a1;">#</span><span style="color: #93a1a1;">' -output psl :             format de sortie psl</span>
   <span style="color: #93a1a1;">#</span><span style="color: #93a1a1;">' reference_reverse.fasta : s&#233;quence de r&#233;f&#233;rence</span>
   <span style="color: #93a1a1;">#</span><span style="color: #93a1a1;">' trimmed.fastq :           s&#233;quence &#224; aligner</span>
   <span style="color: #93a1a1;">#</span><span style="color: #93a1a1;">' output.psl :              fichier de sortie</span>
   ../../bin/ssahaSNP/ssaha2 -output psl ref.fasta trimmed.fastq &gt; output.psl

   <span style="color: #93a1a1;">## </span><span style="color: #93a1a1;">column annotation based on</span>
   <span style="color: #93a1a1;">## </span><span style="color: #93a1a1;">ftp://ftp.sanger.ac.uk/pub/resources/software/ssahasnp/readme,</span>
   <span style="color: #93a1a1;">## </span><span style="color: #93a1a1;">part (6) some further information</span>
   <span style="color: #93a1a1;"># </span><span style="color: #93a1a1;">la premi&#232;re ligne du fichier .dat, afin d'&#234;tre lu dans R</span>
   cat <span style="color: #b58900; font-weight: bold;">\</span>
       &lt;<span style="color: #2aa198;">(</span> <span style="color: #657b83; font-weight: bold;">echo</span> <span style="color: #2aa198;">"match subject_name index_of_subject read_name s_base q_base s_qual q_qual offset_on_</span><span style="color: #d33682;">subject offset_on_read length_of_snp start_match_of_read end_match_of_read match_direction length_of_subject"</span><span style="color: #d33682;"> </span><span style="color: #2aa198;">)</span><span style="color: #d33682;"> </span><span style="color: #d33682; font-weight: bold;">\</span>
       &lt;<span style="color: #2aa198;">(</span> ../../bin/ssahaSNP/ssahaSNP ref.fasta trimmed.fastq | <span style="color: #b58900; font-weight: bold;">\</span>
                egrep ssaha:SNP | <span style="color: #b58900; font-weight: bold;">\</span>
                awk <span style="color: #2aa198;">'{print $1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15}'</span><span style="color: #2aa198;">)</span> <span style="color: #b58900; font-weight: bold;">\</span>
       &gt; snp_calling.dat
</pre>
</div>

<p>
Le script est là : <a href="./src/snp_call.sh">./src/snp_call.sh</a>.
</p>
</div>
</div>

<div id="outline-container-sec-4-6" class="outline-3">
<h3 id="sec-4-6"><span class="section-number-3">4.6</span> création de la table d'identification des clones</h3>
<div class="outline-text-3" id="text-4-6">
<p>
Les spectrogrammes contiennent l'info permettant de relier à un
identifiant de séquence le nom expérimental qu'on lui a attribué.
</p>

<div class="org-src-container">

<pre class="src src-python">  <span style="color: #93a1a1;">#</span><span style="color: #93a1a1;">!/usr/bin/env python</span>

  <span style="color: #859900; font-weight: bold;">from</span> Bio <span style="color: #859900; font-weight: bold;">import</span> SeqIO
  <span style="color: #859900; font-weight: bold;">from</span> glob <span style="color: #859900; font-weight: bold;">import</span> glob
  <span style="color: #859900; font-weight: bold;">from</span> os.path <span style="color: #859900; font-weight: bold;">import</span> basename

  <span style="color: #859900; font-weight: bold;">def</span> <span style="color: #268bd2;">sw_or_ws</span><span style="color: #2aa198;">(</span>mutant_name<span style="color: #2aa198;">)</span>:
      <span style="color: #2aa198;">"""</span>
<span style="color: #2aa198;">      Determine si le mutant est SW ou WS</span>
<span style="color: #2aa198;">      """</span>
      <span style="color: #859900; font-weight: bold;">if</span> <span style="color: #2aa198;">'sw'</span> <span style="color: #859900; font-weight: bold;">in</span> mutant_name:
          <span style="color: #859900; font-weight: bold;">return</span> <span style="color: #2aa198;">'sw'</span>
      <span style="color: #859900; font-weight: bold;">else</span>:
          <span style="color: #859900; font-weight: bold;">return</span> <span style="color: #2aa198;">'ws'</span>

  <span style="color: #93a1a1;"># </span><span style="color: #93a1a1;">en tete de colonne</span>
  <span style="color: #859900; font-weight: bold;">print</span> <span style="color: #2aa198;">"id name mutant"</span>

  <span style="color: #859900; font-weight: bold;">def</span> <span style="color: #268bd2;">print_seq_id</span><span style="color: #2aa198;">(</span>dna_seq<span style="color: #2aa198;">)</span>:
      <span style="color: #859900; font-weight: bold;">with</span> <span style="color: #657b83; font-weight: bold;">open</span><span style="color: #2aa198;">(</span>dna_seq, <span style="color: #2aa198;">"rb"</span><span style="color: #2aa198;">)</span> <span style="color: #859900; font-weight: bold;">as</span> spectro:
          <span style="color: #268bd2;">sequence</span> = <span style="color: #657b83; font-weight: bold;">list</span><span style="color: #2aa198;">(</span>SeqIO.parse<span style="color: #b58900;">(</span>spectro, <span style="color: #2aa198;">"abi"</span><span style="color: #b58900;">)</span><span style="color: #2aa198;">)</span>
          <span style="color: #859900; font-weight: bold;">print</span> sequence<span style="color: #2aa198;">[</span>0<span style="color: #2aa198;">]</span>.<span style="color: #657b83; font-weight: bold;">id</span> + <span style="color: #2aa198;">" "</span> + sequence<span style="color: #2aa198;">[</span>0<span style="color: #2aa198;">]</span>.name + <span style="color: #2aa198;">" "</span> + sw_or_ws<span style="color: #2aa198;">(</span>sequence<span style="color: #b58900;">[</span>0<span style="color: #b58900;">]</span>.name<span style="color: #2aa198;">)</span>

  <span style="color: #859900; font-weight: bold;">for</span> <span style="color: #657b83; font-weight: bold;">file</span> <span style="color: #859900; font-weight: bold;">in</span> glob<span style="color: #2aa198;">(</span><span style="color: #2aa198;">"data/sw/spectro/*.ab1"</span><span style="color: #2aa198;">)</span>:
      <span style="color: #859900; font-weight: bold;">if</span> basename<span style="color: #2aa198;">(</span><span style="color: #657b83; font-weight: bold;">file</span><span style="color: #2aa198;">)</span> != <span style="color: #2aa198;">"psw76-1073bis.ab1"</span>: <span style="color: #93a1a1;"># </span><span style="color: #93a1a1;">exclut la sequence qui pose probleme</span>
          print_seq_id<span style="color: #2aa198;">(</span><span style="color: #657b83; font-weight: bold;">file</span><span style="color: #2aa198;">)</span>

  <span style="color: #859900; font-weight: bold;">for</span> <span style="color: #657b83; font-weight: bold;">file</span> <span style="color: #859900; font-weight: bold;">in</span> glob<span style="color: #2aa198;">(</span><span style="color: #2aa198;">"data/ws/spectro/*.ab1"</span><span style="color: #2aa198;">)</span>:
      print_seq_id<span style="color: #2aa198;">(</span><span style="color: #657b83; font-weight: bold;">file</span><span style="color: #2aa198;">)</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-sec-4-7" class="outline-3">
<h3 id="sec-4-7"><span class="section-number-3">4.7</span> Analyses des résultats de snp calling</h3>
</div>
</div>
<div id="outline-container-sec-5" class="outline-2">
<h2 id="sec-5"><span class="section-number-2">5</span> Tests</h2>
<div class="outline-text-2" id="text-5">
<p>
La fonction <i>assert</i> provient de là :
<a href="http://tldp.org/LDP/abs/html/debugging.html#ASSERT">http://tldp.org/LDP/abs/html/debugging.html#ASSERT</a>
</p>

<p>
Le script en question vient de là : <a href="https://github.com/lehmannro/assert.sh">https://github.com/lehmannro/assert.sh</a>
</p>

<div class="org-src-container">

<pre class="src src-sh">wget https://raw.github.com/lehmannro/assert.sh/v1.1/assert.sh -O test/assert.sh
chmod +x test/assert.sh
</pre>
</div>

<p>
Le script pour tester les résultats, via <code>make test</code>.
</p>

<div class="org-src-container">

<pre class="src src-sh">  <span style="color: #93a1a1;">#</span><span style="color: #93a1a1;">!/bin/bash</span>

  <span style="color: #93a1a1;"># </span><span style="color: #93a1a1;">cette fonction renvoit le nombre de s&#233;quence dans un fichier fastq pool&#233;.</span>
  <span style="color: #93a1a1;"># </span><span style="color: #93a1a1;">elle divise le nombre de ligne par 4 pour conna&#238;tre le nombre de s&#233;quence.</span>
  check_fastq_length <span style="color: #2aa198;">()</span> <span style="color: #2aa198;">{</span>
      cat $<span style="color: #268bd2;">1</span> | <span style="color: #657b83; font-weight: bold;">echo</span> $<span style="color: #b58900;">(</span><span style="color: #268bd2;">(</span><span style="color: #6c71c4; font-weight: bold;">`wc -l`</span> / 4<span style="color: #268bd2;">)</span><span style="color: #b58900;">)</span>
  <span style="color: #2aa198;">}</span>

  <span style="color: #657b83; font-weight: bold;">set</span> -e

  <span style="color: #657b83; font-weight: bold;">.</span> ./test/assert.sh

  <span style="color: #93a1a1;"># </span><span style="color: #93a1a1;">v&#233;rifie qu'il y ait bien 95 s&#233;quence dans le fastq SW et 96 dans le fastq WS</span>
  assert <span style="color: #2aa198;">"echo `check_fastq_length data/sw/sw_untrimmed.fastq`"</span> <span style="color: #2aa198;">"95"</span>
  assert <span style="color: #2aa198;">"echo `check_fastq_length data/ws/ws_untrimmed.fastq`"</span> <span style="color: #2aa198;">"96"</span>

  <span style="color: #93a1a1;"># </span><span style="color: #93a1a1;">v&#233;rifie qu'yl y ait bien 95 s&#233;quences dans le fastq SW trimm&#233; et 96</span>
  <span style="color: #93a1a1;"># </span><span style="color: #93a1a1;">dans le fastq WS trimm&#233;</span>
  assert <span style="color: #2aa198;">"echo `check_fastq_length data/sw/sw_untrimmed.fastq.trim`"</span> <span style="color: #2aa198;">"90"</span>
  assert <span style="color: #2aa198;">"echo `check_fastq_length data/ws/ws_untrimmed.fastq.trim`"</span> <span style="color: #2aa198;">"90"</span>
  assert <span style="color: #2aa198;">"echo `wc -l data/snp_call/snp_calling.dat | awk '{print $1}'`"</span> <span style="color: #2aa198;">"1875"</span>

  assert_end results
</pre>
</div>
</div>
</div>

<div id="outline-container-sec-6" class="outline-2">
<h2 id="sec-6"><span class="section-number-2">6</span> Dépendances</h2>
<div class="outline-text-2" id="text-6">
</div><div id="outline-container-sec-6-1" class="outline-3">
<h3 id="sec-6-1"><span class="section-number-3">6.1</span> BBmap</h3>
<div class="outline-text-3" id="text-6-1">
<p>
Le site de téléchargement est là :
<a href="https://sourceforge.net/projects/bbmap">https://sourceforge.net/projects/bbmap</a>
</p>

<p>
Pour installer bbduk :
</p>

<div class="org-src-container">

<pre class="src src-sh">  wget https://sourceforge.net/projects/bbmap/files/latest/download <span style="color: #b58900; font-weight: bold;">\</span>
       -o bin/bbmap.tar.gz
  <span style="color: #657b83; font-weight: bold;">cd</span> bin
  tar xzf bin/bbmap.tar.gz
</pre>
</div>
</div>
</div>

<div id="outline-container-sec-6-2" class="outline-3">
<h3 id="sec-6-2"><span class="section-number-3">6.2</span> R markdown</h3>
<div class="outline-text-3" id="text-6-2">
<p>
Il faut une version récente du package <code>rmarkdown</code>.
</p>

<div class="org-src-container">

<pre class="src src-R"><span style="color: #268bd2;">install.package</span><span style="color: #2aa198; font-weight: bold;">(</span><span style="color: #2aa198;">'rmarkdown'</span>, type <span style="color: #268bd2; font-weight: bold;">=</span> <span style="color: #2aa198;">"source"</span><span style="color: #2aa198; font-weight: bold;">)</span>
</pre>
</div>
</div>
</div>
</div>
</div><div class="col-md-3"><nav id="table-of-contents">
<div id="text-table-of-contents" class="bs-docs-sidebar">
<ul class="nav">
<li><a href="#sec-1">1. Dossiers et fichiers</a></li>
<li><a href="#sec-2">2. Contrôle qualité</a></li>
<li><a href="#sec-3">3. Analyse rapide du SNP calling de GATC</a></li>
<li><a href="#sec-4">4. SNP calling global</a>
<ul class="nav">
<li><a href="#sec-4-1">4.1. trimming des séquences</a></li>
<li><a href="#sec-4-2">4.2. Contrôle qualité des séquences trimmées</a></li>
<li><a href="#sec-4-3">4.3. Pooling des séquences dans un seul fastq</a></li>
<li><a href="#sec-4-4">4.4. Analyses des alignements via seaview</a></li>
<li><a href="#sec-4-5">4.5. snp calling via ssahaSNP</a></li>
<li><a href="#sec-4-6">4.6. création de la table d'identification des clones</a></li>
<li><a href="#sec-4-7">4.7. Analyses des résultats de snp calling</a></li>
</ul>
</li>
<li><a href="#sec-5">5. Tests</a></li>
<li><a href="#sec-6">6. Dépendances</a>
<ul class="nav">
<li><a href="#sec-6-1">6.1. BBmap</a></li>
<li><a href="#sec-6-2">6.2. R markdown</a></li>
</ul>
</li>
</ul>
</div>
</nav>
</div></div></div>
<footer id="postamble" class="">
<div><p class="author">Auteur: Samuel Barreto</p>
<p class="date">Created: 2016-03-30 Wed 17:38</p>
<p class="creator"><a href="http://www.gnu.org/software/emacs/">Emacs</a> 24.5.1 (<a href="http://orgmode.org">Org-mode</a> 8.3.4)</p>
</div>
</footer>
</body>
</html>
