<!doctype html>
<html lang="fr">
<head>
<title>Séquençage WS et SW</title>
<!-- 2016-04-04 Mon 18:26 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta name="generator" content="Org-mode">
<meta name="author" content="Samuel Barreto">

<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.6/paper/bootstrap.min.css" rel="stylesheet" integrity="sha384-2mX2PSpkRSXLQzmNzH3gwK6srb06+OfbDlYjbog8LQuALYJjuQ3+Yzy2JIWNV9rW" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/1.11.3/jquery.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.5/js/bootstrap.min.js"></script>
<style type="text/css">
/* org mode styles on top of twbs */

html {
    position: relative;
    min-height: 100%;
}

body {
    font-size: 18px;
    margin-bottom: 105px;
}

footer {
    position: absolute;
    bottom: 0;
    width: 100%;
    height: 101px;
    background-color: #f5f5f5;
}

footer > div {
    padding: 10px;
}

footer p {
    margin: 0 0 5px;
    text-align: center;
    font-size: 16px;
}

#table-of-contents {
    margin-top: 20px;
    margin-bottom: 20px;
}

blockquote p {
    font-size: 18px;
}

pre {
    font-size: 16px;
}

.footpara {
    display: inline-block;
}

figcaption {
  font-size: 16px;
  color: #666;
  font-style: italic;
  padding-bottom: 15px;
}

/* from twbs docs */

.bs-docs-sidebar.affix {
    position: static;
}
@media (min-width: 768px) {
    .bs-docs-sidebar {
        padding-left: 20px;
    }
}

/* All levels of nav */
.bs-docs-sidebar .nav > li > a {
    display: block;
    padding: 4px 20px;
    font-size: 14px;
    font-weight: 500;
    color: #999;
}
.bs-docs-sidebar .nav > li > a:hover,
.bs-docs-sidebar .nav > li > a:focus {
    padding-left: 19px;
    color: #A1283B;
    text-decoration: none;
    background-color: transparent;
    border-left: 1px solid #A1283B;
}
.bs-docs-sidebar .nav > .active > a,
.bs-docs-sidebar .nav > .active:hover > a,
.bs-docs-sidebar .nav > .active:focus > a {
    padding-left: 18px;
    font-weight: bold;
    color: #A1283B;
    background-color: transparent;
    border-left: 2px solid #A1283B;
}

/* Nav: second level (shown on .active) */
.bs-docs-sidebar .nav .nav {
    display: none; /* Hide by default, but at >768px, show it */
    padding-bottom: 10px;
}
.bs-docs-sidebar .nav .nav > li > a {
    padding-top: 1px;
    padding-bottom: 1px;
    padding-left: 30px;
    font-size: 12px;
    font-weight: normal;
}
.bs-docs-sidebar .nav .nav > li > a:hover,
.bs-docs-sidebar .nav .nav > li > a:focus {
    padding-left: 29px;
}
.bs-docs-sidebar .nav .nav > .active > a,
.bs-docs-sidebar .nav .nav > .active:hover > a,
.bs-docs-sidebar .nav .nav > .active:focus > a {
    padding-left: 28px;
    font-weight: 500;
}

/* Nav: third level (shown on .active) */
.bs-docs-sidebar .nav .nav .nav {
    padding-bottom: 10px;
}
.bs-docs-sidebar .nav .nav .nav > li > a {
    padding-top: 1px;
    padding-bottom: 1px;
    padding-left: 40px;
    font-size: 12px;
    font-weight: normal;
}
.bs-docs-sidebar .nav .nav .nav > li > a:hover,
.bs-docs-sidebar .nav .nav .nav > li > a:focus {
    padding-left: 39px;
}
.bs-docs-sidebar .nav .nav .nav > .active > a,
.bs-docs-sidebar .nav .nav .nav > .active:hover > a,
.bs-docs-sidebar .nav .nav .nav > .active:focus > a {
    padding-left: 38px;
    font-weight: 500;
}

/* Show and affix the side nav when space allows it */
@media (min-width: 992px) {
    .bs-docs-sidebar .nav > .active > ul {
        display: block;
    }
    /* Widen the fixed sidebar */
    .bs-docs-sidebar.affix,
    .bs-docs-sidebar.affix-bottom {
        width: 213px;
    }
    .bs-docs-sidebar.affix {
        position: fixed; /* Undo the static from mobile first approach */
        top: 20px;
    }
    .bs-docs-sidebar.affix-bottom {
        position: absolute; /* Undo the static from mobile first approach */
    }
    .bs-docs-sidebar.affix .bs-docs-sidenav,.bs-docs-sidebar.affix-bottom .bs-docs-sidenav {
        margin-top: 0;
        margin-bottom: 0
    }
}
@media (min-width: 1200px) {
    /* Widen the fixed sidebar again */
    .bs-docs-sidebar.affix-bottom,
    .bs-docs-sidebar.affix {
        width: 263px;
    }
}
</style>
<script type="text/javascript">
$(function() {
    'use strict';

    $('.bs-docs-sidebar li').first().addClass('active');

    $(document.body).scrollspy({target: '.bs-docs-sidebar'});

    $('.bs-docs-sidebar').affix();
});
</script>
</head>
<body>
<div id="content" class="container">
<div class="row"><div class="col-md-9"><h1 class="title">Séquençage WS et SW</h1>
<p>
Ce dossier contient les séquences des plaques obtenues avec les
constructions SW et WS, envoyées à séquencer le <span class="timestamp-wrapper"><span class="timestamp">[2016-03-15 Tue]</span></span>.
</p>

<div id="outline-container-sec-1" class="outline-2">
<h2 id="sec-1"><span class="section-number-2">1</span> Dossiers et fichiers</h2>
<div class="outline-text-2" id="text-1">
<p>
Les données brutes au format .zip sont dans le dossier <code>raw</code>. Elles
sont extraites dans le dossier <code>data</code>, séparé en <code>ws</code> et <code>sw</code> pour les
analyses. Rien ne doit toucher au dossier <code>raw</code>.
</p>

<p>
J'ai ensuite utilisé le script <a href="./src/sort_into_dir.sh">./src/sort_into_dir.sh</a> pour classer les
fichiers selon leur extension.
</p>

<div class="org-src-container">

<pre class="src src-sh"><span style="color: #8D8D84;">#</span><span style="color: #8D8D84; font-style: italic;">!/bin/</span><span style="color: #0000FF;">bash</span>

<span style="color: #006FE0;">cd</span> data/ws

mkdir fasta spectro seq

find . -name <span style="color: #008000;">"*.fas"</span>  -maxdepth <span style="color: #D0372D;">1</span> -exec mv <span style="color: #707183;">{}</span> ./fasta/ <span style="color: #008000;">\;</span>
find . -name <span style="color: #008000;">"*.ab1"</span>  -maxdepth <span style="color: #D0372D;">1</span> -exec mv <span style="color: #707183;">{}</span> ./spectro/ <span style="color: #008000;">\;</span>
find . -name <span style="color: #008000;">"*.seq"</span>  -maxdepth <span style="color: #D0372D;">1</span> -exec mv <span style="color: #707183;">{}</span> ./seq/ <span style="color: #008000;">\;</span>
find . -name <span style="color: #008000;">"*.csv"</span>  -maxdepth <span style="color: #D0372D;">1</span> -exec mv <span style="color: #707183;">{}</span> ../csv/ <span style="color: #008000;">\;</span>

<span style="color: #006FE0;">cd</span> ../sw

mkdir fasta spectro seq

find . -name <span style="color: #008000;">"*.fas"</span>  -maxdepth <span style="color: #D0372D;">1</span> -exec mv <span style="color: #707183;">{}</span> ./fasta/ <span style="color: #008000;">\;</span>
find . -name <span style="color: #008000;">"*.ab1"</span>  -maxdepth <span style="color: #D0372D;">1</span> -exec mv <span style="color: #707183;">{}</span> ./spectro/ <span style="color: #008000;">\;</span>
find . -name <span style="color: #008000;">"*.seq"</span>  -maxdepth <span style="color: #D0372D;">1</span> -exec mv <span style="color: #707183;">{}</span> ./seq/ <span style="color: #008000;">\;</span>
find . -name <span style="color: #008000;">"*.csv"</span>  -maxdepth <span style="color: #D0372D;">1</span> -exec mv <span style="color: #707183;">{}</span> ../csv/ <span style="color: #008000;">\;</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-sec-2" class="outline-2">
<h2 id="sec-2"><span class="section-number-2">2</span> Contrôle qualité</h2>
<div class="outline-text-2" id="text-2">
<p>
Après avoir reçu les séquences, j'ai voulu connaître la qualité
globale des séquences. Je les ai donc converties en fastq avec le
script <a href="./src/ab1_to_fastq.sh">./src/ab1_to_fastq.sh</a>, et j'ai utilisé <code>fastqc</code> pour
représenter la qualité globale via le script <a href="./src/quality_control.sh">./src/quality_control.sh</a>.
</p>

<div class="org-src-container">

<pre class="src src-sh"><span style="color: #8D8D84;">#</span><span style="color: #8D8D84; font-style: italic;">!/bin/</span><span style="color: #0000FF;">bash</span>

<span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">transforme en fastq</span>
<span style="color: #006FE0;">cd</span> data/ws

touch ws_untrimmed.fastq

<span style="color: #0000FF;">for</span> file<span style="color: #0000FF;"> in</span> spectro/*.ab1
<span style="color: #0000FF;">do</span>
    seqret <span style="color: #008000;">\</span>
        -sformat abi <span style="color: #008000;">\</span>
        -osformat fastq <span style="color: #008000;">\</span>
        -auto <span style="color: #008000;">\</span>
        -stdout <span style="color: #008000;">\</span>
        -sequence $<span style="color: #BA36A5;">file</span> <span style="color: #008000;">\</span>
        &gt;&gt; ws_untrimmed.fastq
<span style="color: #0000FF;">done</span>

<span style="color: #006FE0;">cd</span> ../sw

<span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">TEMPORAIRE, &#233;limine le fichier .ab1 tendancieux pour l'instant</span>
<span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">rm spectro/psw76-1073bis.ab1</span>

touch sw_untrimmed.fastq

<span style="color: #0000FF;">for</span> file<span style="color: #0000FF;"> in</span> spectro/*.ab1
<span style="color: #0000FF;">do</span>
    seqret <span style="color: #008000;">\</span>
        -sformat abi <span style="color: #008000;">\</span>
        -osformat fastq <span style="color: #008000;">\</span>
        -auto <span style="color: #008000;">\</span>
        -stdout <span style="color: #008000;">\</span>
        -sequence $<span style="color: #BA36A5;">file</span> <span style="color: #008000;">\</span>
        &gt;&gt; sw_untrimmed.fastq
<span style="color: #0000FF;">done</span>
</pre>
</div>

<p>
J'ai utilisé <code>fastqc</code> pour représenter la qualité globale via le
script <a href="./src/quality_control.sh">./src/quality_control.sh</a>.
</p>

<div class="org-src-container">

<pre class="src src-sh"><span style="color: #8D8D84;">#</span><span style="color: #8D8D84; font-style: italic;">!/bin/</span><span style="color: #0000FF;">bash</span>

<span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">contr&#244;le qualit&#233; des s&#233;quences.</span>
<span style="color: #006FE0;">cd</span> data

<span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">cr&#233;e un dossier qc de quality control pour ws et sw</span>
mkdir -p qc/<span style="color: #707183;">{</span>ws,sw<span style="color: #707183;">}</span>

<span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">fastqc sortie dasn le dossier appropri&#233;</span>
fastqc ws/ws_untrimmed.fastq -o qc/ws
fastqc sw/sw_untrimmed.fastq -o qc/sw

<span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">transfert du fichier .png qui nous int&#233;resse dans le dir adapt&#233;</span>
<span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">et suppression des dossiers inutiles</span>
<span style="color: #8D8D84;">#</span><span style="color: #8D8D84; font-style: italic;">' -qq : very quiet</span>
<span style="color: #8D8D84;">#</span><span style="color: #8D8D84; font-style: italic;">' -o  : overwrite without warning</span>
<span style="color: #8D8D84;">#</span><span style="color: #8D8D84; font-style: italic;">' -d  : out dir</span>
unzip -qq -o <span style="color: #008000;">\</span>
      qc/sw/sw_untrimmed_fastqc.zip <span style="color: #008000;">\</span>
      sw_untrimmed_fastqc/Images/per_base_quality.png <span style="color: #008000;">\</span>
      -d ../anl/ <span style="color: #008000;">\</span>
    &amp;&amp; mv ../anl/sw_untrimmed_fastqc/Images/per_base_quality.png <span style="color: #008000;">\</span>
          ../anl/sw_per_base_qual.png <span style="color: #008000;">\</span>
    &amp;&amp; rm -r ../anl/sw_untrimmed_fastqc

<span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">transfert du fichier .png qui nous int&#233;resse dans le dir adapt&#233;</span>
<span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">et suppression des dossiers inutiles</span>
unzip -qq -o <span style="color: #008000;">\</span>
      qc/ws/ws_untrimmed_fastqc.zip <span style="color: #008000;">\</span>
      ws_untrimmed_fastqc/Images/per_base_quality.png <span style="color: #008000;">\</span>
      -d ../anl/ <span style="color: #008000;">\</span>
    &amp;&amp; mv ../anl/ws_untrimmed_fastqc/Images/per_base_quality.png <span style="color: #008000;">\</span>
          ../anl/ws_per_base_qual.png <span style="color: #008000;">\</span>
    &amp;&amp; rm -r ../anl/ws_untrimmed_fastqc
</pre>
</div>

<p>
À priori les séquences sont plutôt de bonne qualité. Voir le résumé
dans le fichier <a href="./data/qc/sw/sw_untrimmed_fastqc.html">./data/qc/sw/sw_untrimmed_fastqc.html</a> et
<a href="./data/qc/ws/ws_untrimmed_fastqc.html">./data/qc/ws/ws_untrimmed_fastqc.html</a>.
</p>
</div>
</div>

<div id="outline-container-sec-3" class="outline-2">
<h2 id="sec-3"><span class="section-number-2">3</span> Analyse rapide du SNP calling de GATC</h2>
<div class="outline-text-2" id="text-3">
<p>
J'ai rapidement analysé la sortie du SNP calling effectué par GATC,
mais il manque des données. Je vais refaire le SNP calling moi-même,
en utilisant dans un premier temps ssaha2.
</p>

<p>
Le script est là <a href="./anl/gatc_snpcall/snp_call.r">./anl/gatc_snpcall/snp_call.r</a>, les sorties en
html là <a href="./anl/gatc_snpcall/snp_call.html">./anl/gatc_snpcall/snp_call.html</a>.
</p>
</div>
</div>

<div id="outline-container-sec-4" class="outline-2">
<h2 id="sec-4"><span class="section-number-2">4</span> SNP calling global</h2>
<div class="outline-text-2" id="text-4">
</div><div id="outline-container-sec-4-1" class="outline-3">
<h3 id="sec-4-1"><span class="section-number-3">4.1</span> trimming des séquences</h3>
<div class="outline-text-3" id="text-4-1">
<p>
J'ai utilisé le programme bbduk pour trimmer les séquences de faible
qualité.
</p>

<div class="org-src-container">

<pre class="src src-sh">  <span style="color: #8D8D84;">#</span><span style="color: #8D8D84; font-style: italic;">!/bin/bash</span>

  <span style="color: #8D8D84;">#</span><span style="color: #8D8D84; font-style: italic;">' -qtrim=rl : quality trim right and left</span>
  <span style="color: #8D8D84;">#</span><span style="color: #8D8D84; font-style: italic;">' -trimq=28 : trim if quality &lt; 28 (sanger encoding, illumina 1.9)</span>
  <span style="color: #8D8D84;">#</span><span style="color: #8D8D84; font-style: italic;">' -minlen=620 : keep only seq with length &gt; 620, after trimming.</span>
  <span style="color: #8D8D84;">#</span><span style="color: #8D8D84; font-style: italic;">' -Xmx1g : tells bbduk / java to use 1G of RAM</span>

  <span style="color: #BA36A5;">FASTQ</span>=<span style="color: #008000;">"data/sw/sw_untrimmed.fastq</span>
<span style="color: #008000;">  data/ws/ws_untrimmed.fastq</span>
<span style="color: #008000;">  "</span>

  <span style="color: #0000FF;">for</span> f<span style="color: #0000FF;"> in</span> $<span style="color: #BA36A5;">FASTQ</span>
  <span style="color: #0000FF;">do</span>
      <span style="color: #BA36A5;">out</span>=$<span style="color: #BA36A5;">f</span>.trim
      bin/bbmap/bbduk.sh -Xmx1g <span style="color: #008000;">\</span>
                         -in=$<span style="color: #BA36A5;">f</span> <span style="color: #008000;">\</span>
                         -out=$<span style="color: #BA36A5;">f</span>.qtrim <span style="color: #008000;">\</span>
                         -qtrim=rl <span style="color: #008000;">\</span>
                         -trimq=<span style="color: #D0372D;">28</span> <span style="color: #008000;">\</span>
                         -minlen=<span style="color: #D0372D;">600</span>

      <span style="color: #8D8D84;">## </span><span style="color: #8D8D84; font-style: italic;">convertit les bases d'une qualit&#233; inf&#233;rieure &#224; 20 en N.</span>
      seqtk seq <span style="color: #008000;">\</span>
            -q20 <span style="color: #008000;">\</span>
            -nN <span style="color: #008000;">\</span>
            $<span style="color: #BA36A5;">f</span>.qtrim &gt; $<span style="color: #BA36A5;">out</span>

      <span style="color: #8D8D84;">## </span><span style="color: #8D8D84; font-style: italic;">convertit le fastq en fasta</span>
      seqret <span style="color: #008000;">\</span>
          -sformat fastq <span style="color: #008000;">\</span>
          -osformat fasta <span style="color: #008000;">\</span>
          -auto -stdout <span style="color: #008000;">\</span>
          -sequence $<span style="color: #BA36A5;">out</span> &gt; $<span style="color: #BA36A5;">f</span>.fasta

      rm $<span style="color: #BA36A5;">f</span>.qtrim
  <span style="color: #0000FF;">done</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-sec-4-2" class="outline-3">
<h3 id="sec-4-2"><span class="section-number-3">4.2</span> Contrôle qualité des séquences trimmées</h3>
<div class="outline-text-3" id="text-4-2">
<div class="org-src-container">

<pre class="src src-sh">  <span style="color: #8D8D84;">#</span><span style="color: #8D8D84; font-style: italic;">!/usr/bin/env bash</span>

  <span style="color: #006FE0;">cd</span> data

  <span style="color: #BA36A5;">TRIM</span>=<span style="color: #008000;">"ws/ws_untrimmed.fastq.trim sw/sw_untrimmed.fastq.trim"</span>

  mkdir qtrim_qc

  <span style="color: #0000FF;">for</span> f<span style="color: #0000FF;"> in</span> $<span style="color: #BA36A5;">TRIM</span>
  <span style="color: #0000FF;">do</span>
      fastqc $<span style="color: #BA36A5;">f</span> -o qtrim_qc
  <span style="color: #0000FF;">done</span>
</pre>
</div>

<p>
Le script est là <a href="./src/qtrim_qc.sh">./src/qtrim_qc.sh</a> et les sorties sont là
<a href="data/qtrim_qc/sw_untrimmed.fastq.trim_fastqc.html">./data/qtrim<sub>qc</sub>/sw<sub>untrimmed.fastq.trim</sub><sub>fastqc.html</sub></a> et là
<a href="data/qtrim_qc/sw_untrimmed.fastq.trim_fastqc.html">./data/qtrim<sub>qc</sub>/ws<sub>untrimmed.fastq.trim</sub><sub>fastqc.html</sub></a>.
</p>

<p>
Les séquences sont nettement raccourcies par rapport aux séquences non trimmées. Mais l'ensemble des
bases qu'on veut analyser sont de bonne qualité. C'est donc plutôt bon signe.
</p>
</div>
</div>

<div id="outline-container-sec-4-3" class="outline-3">
<h3 id="sec-4-3"><span class="section-number-3">4.3</span> Pooling des séquences dans un seul fastq</h3>
<div class="outline-text-3" id="text-4-3">
<div class="org-src-container">

<pre class="src src-sh">  <span style="color: #8D8D84;">#</span><span style="color: #8D8D84; font-style: italic;">!/usr/bin/env bash</span>

  cat data/sw/sw_untrimmed.fastq.trim data/ws/ws_untrimmed.fastq.trim &gt; data/trimmed.fastq
</pre>
</div>

<p>
<a href="./src/pool_trim.sh">./src/pool_trim.sh</a>
</p>
</div>
</div>

<div id="outline-container-sec-4-4" class="outline-3">
<h3 id="sec-4-4"><span class="section-number-3">4.4</span> Analyses des alignements via seaview</h3>
<div class="outline-text-3" id="text-4-4">
<p>
Pour les analyses dans seaview, il faut un fichier fasta. Je lui
ajoute la référence.
</p>

<div class="org-src-container">

<pre class="src src-sh"><span style="color: #8D8D84;">#</span><span style="color: #8D8D84; font-style: italic;">!/usr/bin/</span><span style="color: #0000FF;">env</span><span style="color: #8D8D84; font-style: italic;"> bash</span>

seqret -sformat fastq -osformat fasta -auto -stdout <span style="color: #008000;">\</span>
       -sequence data/trimmed.fastq &gt; data/trimmed.fasta

fastx_reverse_complement <span style="color: #008000;">\</span>
    -i raw/ref/reference1073bis-1392.fasta <span style="color: #008000;">\</span>
    &gt;&gt; data/trimmed.fasta
</pre>
</div>

<p>
<a href="./src/trim_to_fasta.sh">./src/trim_to_fasta.sh</a>
</p>

<p>
J'ai placé les différentes analyses graphiques via seaview dans le
dossier <a href="./anl/align">./anl/align</a>.
</p>

<p>
Différents alignements :
</p>
<ul class="org-ul">
<li>l'alignement global : <a href="./anl/align/trimmed.mase">./anl/align/trimmed.mase</a>
</li>
<li>l'alignement avec seulement les sites variables : <a href="./anl/align/variable_site.mase">./anl/align/variable_site.mase</a>
</li>
</ul>

<p>
Sur l'alignement global, j'ai repéré quelques cas de traces complexes,
où des séquences avaient ‘gardé’ leur allèle sauvage, quand les deux
SNP environnants sont les SNP transformants.
</p>
</div>
</div>

<div id="outline-container-sec-4-5" class="outline-3">
<h3 id="sec-4-5"><span class="section-number-3">4.5</span> snp calling via ssahaSNP</h3>
<div class="outline-text-3" id="text-4-5">
<p>
Pour installer ssahaSNP voir les instructions là :
<a href="ftp://ftp.sanger.ac.uk/pub/resources/software/ssahasnp/">ftp://ftp.sanger.ac.uk/pub/resources/software/ssahasnp/</a>.
</p>

<p>
Il est très important que les fichiers fasta de référence soit formattés de la
bonne façon, sinon ssahaSNP ne sort pas les bonnes valeurs dans les lignes de
tableau.
</p>

<div class="org-src-container">

<pre class="src src-sh">   <span style="color: #8D8D84;">#</span><span style="color: #8D8D84; font-style: italic;">!/usr/bin/env bash</span>

   <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">Variant calling using ssaha2 and ssahaSNP</span>

   <span style="color: #0000FF;">if</span> <span style="color: #707183;">[</span> ! -e data/snp_call <span style="color: #707183;">]</span>
   <span style="color: #0000FF;">then</span>
       mkdir data/snp_call
   <span style="color: #0000FF;">fi</span>

   <span style="color: #006FE0;">cd</span> data/snp_call

   <span style="color: #0000FF;">if</span> <span style="color: #707183;">[</span><span style="color: #7388D6;">[</span> ! -e trimmed.fastq &amp;&amp; ! -e ref.fasta <span style="color: #7388D6;">]</span><span style="color: #707183;">]</span>
   <span style="color: #0000FF;">then</span>
       ln -s    ../trimmed.fastq .
       ln -s ../../raw/ref/reference1073bis-1392.fasta ref.fasta
   <span style="color: #0000FF;">fi</span>

   <span style="color: #8D8D84;">## </span><span style="color: #8D8D84; font-style: italic;">alignement &#224; la s&#233;quence de r&#233;f&#233;rence</span>
   <span style="color: #8D8D84;">#</span><span style="color: #8D8D84; font-style: italic;">' -output psl :             format de sortie psl</span>
   <span style="color: #8D8D84;">#</span><span style="color: #8D8D84; font-style: italic;">' reference_reverse.fasta : s&#233;quence de r&#233;f&#233;rence</span>
   <span style="color: #8D8D84;">#</span><span style="color: #8D8D84; font-style: italic;">' trimmed.fastq :           s&#233;quence &#224; aligner</span>
   <span style="color: #8D8D84;">#</span><span style="color: #8D8D84; font-style: italic;">' output.psl :              fichier de sortie</span>
   ../../bin/ssahaSNP/ssaha2 -output psl ref.fasta trimmed.fastq &gt; output.psl

   <span style="color: #8D8D84;">## </span><span style="color: #8D8D84; font-style: italic;">column annotation based on</span>
   <span style="color: #8D8D84;">## </span><span style="color: #8D8D84; font-style: italic;">ftp://ftp.sanger.ac.uk/pub/resources/software/ssahasnp/readme,</span>
   <span style="color: #8D8D84;">## </span><span style="color: #8D8D84; font-style: italic;">part (6) some further information</span>
   <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">la premi&#232;re ligne du fichier .dat, afin d'&#234;tre lu dans R</span>
   cat <span style="color: #008000;">\</span>
       &lt;<span style="color: #707183;">(</span> <span style="color: #006FE0;">echo</span> <span style="color: #008000;">"match subject_name index_of_subject read_name s_base q_base s_qual q_qual offset_on_subject offset_on_read length_of_snp start_match_of_read end_match_of_read match_direction length_of_subject"</span> <span style="color: #707183;">)</span> <span style="color: #008000;">\</span>
       &lt;<span style="color: #707183;">(</span> ../../bin/ssahaSNP/ssahaSNP ref.fasta trimmed.fastq | <span style="color: #008000;">\</span>
                egrep ssaha:SNP | <span style="color: #008000;">\</span>
                awk <span style="color: #008000;">'{print $1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15}'</span><span style="color: #707183;">)</span> <span style="color: #008000;">\</span>
       &gt; snp_calling.dat
</pre>
</div>

<p>
Le script est là : <a href="./src/snp_call.sh">./src/snp_call.sh</a>.
</p>
</div>
</div>

<div id="outline-container-sec-4-6" class="outline-3">
<h3 id="sec-4-6"><span class="section-number-3">4.6</span> création de la table d'identification des clones</h3>
<div class="outline-text-3" id="text-4-6">
<p>
Les spectrogrammes contiennent l'info permettant de relier à un
identifiant de séquence le nom expérimental qu'on lui a attribué.
</p>

<div class="org-src-container">

<pre class="src src-python">  <span style="color: #8D8D84;">#</span><span style="color: #8D8D84; font-style: italic;">!/usr/bin/env python</span>

  <span style="color: #0000FF;">from</span> Bio <span style="color: #0000FF;">import</span> SeqIO
  <span style="color: #0000FF;">from</span> glob <span style="color: #0000FF;">import</span> glob
  <span style="color: #0000FF;">from</span> os.path <span style="color: #0000FF;">import</span> basename

  <span style="color: #0000FF;">def</span> <span style="color: #006699;">sw_or_ws</span><span style="color: #707183;">(</span>mutant_name<span style="color: #707183;">)</span>:
      <span style="color: #008000;">"""</span>
<span style="color: #008000;">      Determine si le mutant est SW ou WS</span>
<span style="color: #008000;">      """</span>
      <span style="color: #0000FF;">if</span> <span style="color: #008000;">'sw'</span> <span style="color: #0000FF;">in</span> mutant_name:
          <span style="color: #0000FF;">return</span> <span style="color: #008000;">'sw'</span>
      <span style="color: #0000FF;">else</span>:
          <span style="color: #0000FF;">return</span> <span style="color: #008000;">'ws'</span>

  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">en tete de colonne</span>
  <span style="color: #0000FF;">print</span> <span style="color: #008000;">"id name mutant"</span>

  <span style="color: #0000FF;">def</span> <span style="color: #006699;">print_seq_id</span><span style="color: #707183;">(</span>dna_seq<span style="color: #707183;">)</span>:
      <span style="color: #0000FF;">with</span> <span style="color: #006FE0;">open</span><span style="color: #707183;">(</span>dna_seq, <span style="color: #008000;">"rb"</span><span style="color: #707183;">)</span> <span style="color: #0000FF;">as</span> spectro:
          <span style="color: #BA36A5;">sequence</span> = <span style="color: #006FE0;">list</span><span style="color: #707183;">(</span>SeqIO.parse<span style="color: #7388D6;">(</span>spectro, <span style="color: #008000;">"abi"</span><span style="color: #7388D6;">)</span><span style="color: #707183;">)</span>
          <span style="color: #0000FF;">print</span> sequence<span style="color: #707183;">[</span><span style="color: #D0372D;">0</span><span style="color: #707183;">]</span>.<span style="color: #006FE0;">id</span> + <span style="color: #008000;">" "</span> + sequence<span style="color: #707183;">[</span><span style="color: #D0372D;">0</span><span style="color: #707183;">]</span>.name + <span style="color: #008000;">" "</span> + sw_or_ws<span style="color: #707183;">(</span>sequence<span style="color: #7388D6;">[</span><span style="color: #D0372D;">0</span><span style="color: #7388D6;">]</span>.name<span style="color: #707183;">)</span>

  <span style="color: #0000FF;">for</span> <span style="color: #006FE0;">file</span> <span style="color: #0000FF;">in</span> glob<span style="color: #707183;">(</span><span style="color: #008000;">"data/sw/spectro/*.ab1"</span><span style="color: #707183;">)</span>:
      <span style="color: #0000FF;">if</span> basename<span style="color: #707183;">(</span><span style="color: #006FE0;">file</span><span style="color: #707183;">)</span> != <span style="color: #008000;">"psw76-1073bis.ab1"</span>: <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">exclut la sequence qui pose probleme</span>
          print_seq_id<span style="color: #707183;">(</span><span style="color: #006FE0;">file</span><span style="color: #707183;">)</span>

  <span style="color: #0000FF;">for</span> <span style="color: #006FE0;">file</span> <span style="color: #0000FF;">in</span> glob<span style="color: #707183;">(</span><span style="color: #008000;">"data/ws/spectro/*.ab1"</span><span style="color: #707183;">)</span>:
      print_seq_id<span style="color: #707183;">(</span><span style="color: #006FE0;">file</span><span style="color: #707183;">)</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-sec-4-7" class="outline-3">
<h3 id="sec-4-7"><span class="section-number-3">4.7</span> Analyses des résultats de snp calling</h3>
</div>
</div>
<div id="outline-container-sec-5" class="outline-2">
<h2 id="sec-5"><span class="section-number-2">5</span> SNP calling précis <code>[0%]</code></h2>
<div class="outline-text-2" id="text-5">
<p>
Le SNP calling via ssahaSNP ne permet pas de connaître l'ensemble des
informations dont on a besoin. Un peu dans l'esprit de polySNP, je
veux écrire un script python qui permette :
</p>

<ol class="org-ol">
<li>de lancer phred sur la séquence abi en entrée.
</li>
<li>de lancer muscle avec pour entrée le fichier d'alignement fst entre
le gène synthétique et la séquence sauvage d'une part, et la
séquence dans le fichier seq d'autre part.
</li>
<li>d'analyser la sortie de phred .phd pour extraire l'information de
la qualité de la base.
</li>
</ol>

<p>
Étant donné le nombre de séquence faible, j'ai fait le choix de lancer
muscle à chaque séquence. Une façon plus efficace de faire la même
chose serait d'aligner l'ensemble des séquences SNP WS d'un coup, avec
leur référence, mais c'est un peu plus compliqué à programmer.
</p>

<div class="org-src-container">

<pre class="src src-python"><span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">-*- coding: utf-8 -*-</span>

<span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">1. lancer phred sur la s&#233;quence abi en entr&#233;e.</span>
<span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">2. lancer muscle avec pour entr&#233;e le fichier d'alignement fst entre</span>
<span style="color: #8D8D84;">#    </span><span style="color: #8D8D84; font-style: italic;">le g&#232;ne synth&#233;tique et la s&#233;quence sauvage d'une part, et la</span>
<span style="color: #8D8D84;">#    </span><span style="color: #8D8D84; font-style: italic;">s&#233;quence dans le fichier seq d'autre part.</span>
<span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">3. analyser la sortie de phred .phd pour extraire l'information de</span>
<span style="color: #8D8D84;">#    </span><span style="color: #8D8D84; font-style: italic;">la qualit&#233; de la base.</span>

<span style="color: #0000FF;">from</span> Bio <span style="color: #0000FF;">import</span> AlignIO  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">permet d'analyser les sorties de muscle.</span>
<span style="color: #0000FF;">from</span> Bio <span style="color: #0000FF;">import</span> SeqIO  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">permet de tester les entr&#233;es</span>
<span style="color: #0000FF;">import</span> subprocess <span style="color: #0000FF;">as</span> sub  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">permet de lancer phred et muscle</span>
<span style="color: #0000FF;">import</span> click
<span style="color: #0000FF;">import</span> os.path
<span style="color: #0000FF;">import</span> numpy <span style="color: #0000FF;">as</span> np
<span style="color: #0000FF;">import</span> pandas <span style="color: #0000FF;">as</span> pd


<span style="color: #0000FF;">def</span> <span style="color: #006699;">is_abi</span><span style="color: #707183;">(</span>sequence<span style="color: #707183;">)</span>:
    <span style="color: #008000;">"""Cette fonction d&#233;termine si une s&#233;quence est bien un fichier spectrogramme.</span>
<span style="color: #008000;">    Renvoit TRUE ou FALSE."""</span>
    <span style="color: #0000FF;">assert</span> os.path.isfile<span style="color: #707183;">(</span>sequence<span style="color: #707183;">)</span>, <span style="color: #008000;">"Le fichier n'existe pas."</span>

    <span style="color: #0000FF;">with</span> <span style="color: #006FE0;">open</span><span style="color: #707183;">(</span>sequence,
              <span style="color: #008000;">"rb"</span><span style="color: #707183;">)</span> <span style="color: #0000FF;">as</span> <span style="color: #006FE0;">input</span>:  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">le fichier doit &#234;tre ouvert en mode "rb"</span>
        <span style="color: #0000FF;">if</span> <span style="color: #006FE0;">len</span><span style="color: #707183;">(</span>SeqIO.read<span style="color: #7388D6;">(</span>
                <span style="color: #006FE0;">input</span>,
                <span style="color: #008000;">"abi"</span><span style="color: #7388D6;">)</span>.<span style="color: #006FE0;">id</span><span style="color: #707183;">)</span> &gt; <span style="color: #D0372D;">0</span>:  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">teste la longueur de l'id de la s&#233;quence</span>
            <span style="color: #0000FF;">return</span> <span style="color: #D0372D;">True</span>
        <span style="color: #0000FF;">else</span>:
            <span style="color: #0000FF;">return</span> <span style="color: #D0372D;">False</span>


<span style="color: #0000FF;">def</span> <span style="color: #006699;">is_fasta</span><span style="color: #707183;">(</span>sequence<span style="color: #707183;">)</span>:
    <span style="color: #008000;">"""Cette fonction d&#233;termine si une s&#233;quence est bien un fichier fasta.</span>
<span style="color: #008000;">    Renvoit True ou False."""</span>
    <span style="color: #0000FF;">assert</span> os.path.isfile<span style="color: #707183;">(</span>sequence<span style="color: #707183;">)</span>, <span style="color: #008000;">"Le fichier n'existe pas"</span>

    <span style="color: #0000FF;">with</span> <span style="color: #006FE0;">open</span><span style="color: #707183;">(</span>sequence, <span style="color: #008000;">"rb"</span><span style="color: #707183;">)</span> <span style="color: #0000FF;">as</span> <span style="color: #006FE0;">input</span>:
        <span style="color: #BA36A5;">seq_list</span> = <span style="color: #707183;">[]</span>
        <span style="color: #0000FF;">for</span> record <span style="color: #0000FF;">in</span> SeqIO.parse<span style="color: #707183;">(</span><span style="color: #006FE0;">input</span>, <span style="color: #008000;">"fasta"</span><span style="color: #707183;">)</span>:
            seq_list.append<span style="color: #707183;">(</span><span style="color: #006FE0;">len</span><span style="color: #7388D6;">(</span>record.<span style="color: #006FE0;">id</span><span style="color: #7388D6;">)</span><span style="color: #707183;">)</span>  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">cr&#233;e une liste de longueur d'id.</span>

    <span style="color: #0000FF;">if</span> <span style="color: #D0372D;">0</span> <span style="color: #0000FF;">not</span> <span style="color: #0000FF;">in</span> seq_list:  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">s'il n'y a pas d'identifiant nul</span>
        <span style="color: #0000FF;">return</span> <span style="color: #D0372D;">True</span>  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">c'est bon.</span>
    <span style="color: #0000FF;">else</span>:
        <span style="color: #0000FF;">return</span> <span style="color: #D0372D;">False</span>


<span style="color: #0000FF;">def</span> <span style="color: #006699;">run_phred</span><span style="color: #707183;">(</span><span style="color: #006FE0;">input</span>, cutoff<span style="color: #707183;">)</span>:
    <span style="color: #008000;">"""Cette fonction utilise phred sur la s&#233;quence .abi en entr&#233;e.</span>
<span style="color: #008000;">    Test: entr&#233;e de type abi."""</span>
    <span style="color: #0000FF;">assert</span> is_abi<span style="color: #707183;">(</span><span style="color: #006FE0;">input</span><span style="color: #707183;">)</span>, <span style="color: #008000;">"La s&#233;quence n'est pas un fichier de sequen&#231;age."</span>

    <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">phred reads and process file listed in the file input via `-if`.</span>
    <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">we must then create a temporary file containing only the name of the input.</span>
    <span style="color: #0000FF;">with</span> <span style="color: #006FE0;">open</span><span style="color: #707183;">(</span><span style="color: #008000;">'tempfile'</span>, <span style="color: #008000;">'w'</span><span style="color: #707183;">)</span> <span style="color: #0000FF;">as</span> tempfile:
        tempfile.write<span style="color: #707183;">(</span><span style="color: #006FE0;">input</span> + <span style="color: #008000;">"\n"</span><span style="color: #707183;">)</span>

    <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">subprocess.call takes a list of argument as input.</span>
    <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">the following list describes the phred list of argument.</span>
    <span style="color: #BA36A5;">phred</span> = <span style="color: #707183;">[</span>
        <span style="color: #008000;">"phred"</span>,
        <span style="color: #008000;">"-st"</span>,
        <span style="color: #008000;">"fasta"</span>,  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">sequence type output (default = fasta)</span>
        <span style="color: #008000;">"-trim_alt"</span>,
        <span style="color: #008000;">"\"\""</span>,  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">quality trim</span>
        <span style="color: #008000;">"-trim_cutoff"</span>,
        <span style="color: #006FE0;">str</span><span style="color: #7388D6;">(</span>cutoff<span style="color: #7388D6;">)</span>,  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">error probability of trimming</span>
        <span style="color: #008000;">"-trim_fasta"</span>,  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">trim sortie fasta.</span>
        <span style="color: #008000;">"-trim_phd"</span>,  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">trim sortie phd</span>
        <span style="color: #008000;">"-s"</span>,  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">write seq file, append ".seq" to the names of the input files</span>
        <span style="color: #008000;">"-d"</span>,  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">write a poly file</span>
        <span style="color: #008000;">"-p"</span>,  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">write a phd file</span>
        <span style="color: #008000;">"-if"</span>,
        <span style="color: #008000;">"tempfile"</span>  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">input file</span>
    <span style="color: #707183;">]</span>
    <span style="color: #BA36A5;">phred_call</span> = sub.call<span style="color: #707183;">(</span>phred<span style="color: #707183;">)</span>
    os.remove<span style="color: #707183;">(</span><span style="color: #008000;">"tempfile"</span><span style="color: #707183;">)</span>  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">supprime le fichier temporaire</span>
    <span style="color: #0000FF;">return</span> phred_call


<span style="color: #0000FF;">def</span> <span style="color: #006699;">run_muscle</span><span style="color: #707183;">(</span><span style="color: #006FE0;">input</span>, reference<span style="color: #707183;">)</span>:
    <span style="color: #008000;">"""Cette fonction lance muscle sur la s&#233;quence fasta en entr&#233;e.</span>
<span style="color: #008000;">    Test: entr&#233;e de type fasta."""</span>
    <span style="color: #0000FF;">assert</span> is_fasta<span style="color: #707183;">(</span>reference<span style="color: #707183;">)</span>, <span style="color: #008000;">"La r&#233;f&#233;rence n'est pas un fichier fasta."</span>
    <span style="color: #0000FF;">assert</span> is_fasta<span style="color: #707183;">(</span><span style="color: #006FE0;">input</span><span style="color: #707183;">)</span>, <span style="color: #008000;">"La s&#233;quence n'est pas un fichier fasta."</span>

    <span style="color: #BA36A5;">muscle</span> = <span style="color: #707183;">[</span>
        <span style="color: #008000;">"muscle"</span>,
        <span style="color: #008000;">"-quiet"</span>,  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">non verbose</span>
        <span style="color: #008000;">"-profile"</span>,  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">do not disrupt the profile alignment</span>
        <span style="color: #008000;">"-objscore"</span>,
        <span style="color: #008000;">"ps"</span>,  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">matrice de scoring</span>
        <span style="color: #008000;">"-maxmb"</span>,
        <span style="color: #006FE0;">str</span><span style="color: #7388D6;">(</span><span style="color: #D0372D;">100</span><span style="color: #7388D6;">)</span>,  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">1.5 G as max memory</span>
        <span style="color: #008000;">"-in1"</span>,
        <span style="color: #006FE0;">input</span>,
        <span style="color: #008000;">"-in2"</span>,
        <span style="color: #006FE0;">str</span><span style="color: #7388D6;">(</span>reference<span style="color: #7388D6;">)</span>,  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">profile is reference</span>
        <span style="color: #008000;">"-out"</span>,
        <span style="color: #006FE0;">input</span> + <span style="color: #008000;">".aln"</span>  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">append .aln to input name</span>
    <span style="color: #707183;">]</span>

    <span style="color: #BA36A5;">muscle_call</span> = sub.call<span style="color: #707183;">(</span>muscle<span style="color: #707183;">)</span>  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">call muscle on input</span>
    <span style="color: #0000FF;">return</span> muscle_call  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">return 0 or 1 if muscle is successful.</span>


<span style="color: #6434A3;">@click.group</span><span style="color: #707183;">()</span>
<span style="color: #0000FF;">def</span> <span style="color: #006699;">phruscle</span><span style="color: #707183;">()</span>:
    <span style="color: #008000;">"""This program gather a set of function to run phred on a abi file, to run muscle on the</span>
<span style="color: #008000;">    called bases, and to give a nicely formatted output either as human readable or as csv.</span>

<span style="color: #008000;">    """</span>
    <span style="color: #0000FF;">pass</span>


<span style="color: #6434A3;">@phruscle.command</span><span style="color: #707183;">(</span><span style="color: #008000;">'basecall'</span>, short_help=<span style="color: #008000;">'Make basecall and alignment.'</span><span style="color: #707183;">)</span>
<span style="color: #6434A3;">@click.option</span><span style="color: #707183;">(</span><span style="color: #008000;">'-c'</span>,
              <span style="color: #008000;">'--cutoff'</span>,
              default=<span style="color: #D0372D;">0</span>.<span style="color: #D0372D;">05</span>,
              <span style="color: #006FE0;">help</span>=<span style="color: #008000;">"Probability of error of basecalling"</span><span style="color: #707183;">)</span>
<span style="color: #6434A3;">@click.option</span><span style="color: #707183;">(</span><span style="color: #008000;">'-i'</span>, <span style="color: #008000;">'--input'</span>, <span style="color: #006FE0;">help</span>=<span style="color: #008000;">"The abi sequence to base call"</span><span style="color: #707183;">)</span>
<span style="color: #6434A3;">@click.option</span><span style="color: #707183;">(</span><span style="color: #008000;">'-r'</span>, <span style="color: #008000;">'--ref'</span>, <span style="color: #006FE0;">help</span>=<span style="color: #008000;">"The reference alignment in fasta format"</span><span style="color: #707183;">)</span>
<span style="color: #0000FF;">def</span> <span style="color: #006699;">run_phruscle</span><span style="color: #707183;">(</span><span style="color: #006FE0;">input</span>, ref, cutoff<span style="color: #707183;">)</span>:
    <span style="color: #008000;">"""This programme uses phred to call bases on the `--input` file to create a fasta file.</span>
<span style="color: #008000;">    Muscle align this fasta file to the `--reference`, without disrupting the reference,</span>
<span style="color: #008000;">    to get the localisation of SNPs and their polarity.</span>

<span style="color: #008000;">    """</span>
    <span style="color: #BA36A5;">phred_code</span> = run_phred<span style="color: #707183;">(</span><span style="color: #006FE0;">input</span>, cutoff<span style="color: #707183;">)</span>
    <span style="color: #0000FF;">if</span> phred_code == <span style="color: #D0372D;">0</span>:
        run_muscle<span style="color: #707183;">(</span><span style="color: #006FE0;">input</span> + <span style="color: #008000;">".seq"</span>, ref<span style="color: #707183;">)</span>
    <span style="color: #0000FF;">else</span>:
        <span style="color: #0000FF;">print</span> <span style="color: #008000;">"Phred failed."</span>


<span style="color: #0000FF;">def</span> <span style="color: #006699;">parse_muscle</span><span style="color: #707183;">(</span>align<span style="color: #707183;">)</span>:
    <span style="color: #008000;">"""Cette fonction renvoit un dict comprenant 3 choses :</span>
<span style="color: #008000;">    1. la s&#233;quence align&#233;e</span>
<span style="color: #008000;">    2. la s&#233;quence sauvage</span>
<span style="color: #008000;">    3. la s&#233;quence avec les SNP</span>
<span style="color: #008000;">    """</span>
    <span style="color: #0000FF;">assert</span> is_fasta<span style="color: #707183;">(</span>align<span style="color: #707183;">)</span>, <span style="color: #008000;">"Le fichier n'est pas un fichier fasta"</span>

    <span style="color: #0000FF;">def</span> <span style="color: #006699;">read_seq</span><span style="color: #707183;">(</span>sequence, index<span style="color: #707183;">)</span>:
        <span style="color: #0000FF;">return</span> AlignIO.read<span style="color: #707183;">(</span>sequence, <span style="color: #008000;">"fasta"</span><span style="color: #707183;">)[</span>index<span style="color: #707183;">]</span>

    <span style="color: #0000FF;">return</span> <span style="color: #707183;">{</span>
        <span style="color: #008000;">'exp'</span>: <span style="color: #006FE0;">str</span><span style="color: #7388D6;">(</span>read_seq<span style="color: #909183;">(</span>align, <span style="color: #D0372D;">0</span><span style="color: #909183;">)</span>.seq<span style="color: #7388D6;">)</span>,  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">la s&#233;quence exp&#233;rimentale</span>
        <span style="color: #008000;">'wt'</span>: <span style="color: #006FE0;">str</span><span style="color: #7388D6;">(</span>read_seq<span style="color: #909183;">(</span>align, <span style="color: #D0372D;">1</span><span style="color: #909183;">)</span>.seq<span style="color: #7388D6;">)</span>,  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">la s&#233;quence sauvage.</span>
        <span style="color: #008000;">'snp'</span>: <span style="color: #006FE0;">str</span><span style="color: #7388D6;">(</span>read_seq<span style="color: #909183;">(</span>align, <span style="color: #D0372D;">2</span><span style="color: #909183;">)</span>.seq<span style="color: #7388D6;">)</span>  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">la s&#233;quence du g&#232;ne synth&#233;tique</span>
    <span style="color: #707183;">}</span>


<span style="color: #0000FF;">def</span> <span style="color: #006699;">are_the_same</span><span style="color: #707183;">(</span>exp, snp, wt<span style="color: #707183;">)</span>:
    <span style="color: #008000;">"""D&#233;termine si les trois bases sont identiques. Renvoit `.` quand les trois bases sont</span>
<span style="color: #008000;">    identiques, `x` quand la base exp est la base `snp`, `X` quand la base exp est la base</span>
<span style="color: #008000;">    wt, et '-' quand la base exp ne correspond &#224; aucune des possibilit&#233;s</span>

<span style="color: #008000;">    """</span>
    <span style="color: #0000FF;">if</span> exp == <span style="color: #008000;">'-'</span>:
        <span style="color: #0000FF;">return</span> <span style="color: #008000;">'-'</span>
    <span style="color: #0000FF;">elif</span> exp == snp <span style="color: #0000FF;">and</span> snp == wt:
        <span style="color: #0000FF;">return</span> <span style="color: #008000;">'.'</span>
    <span style="color: #0000FF;">elif</span> exp == snp <span style="color: #0000FF;">and</span> snp != wt:
        <span style="color: #0000FF;">return</span> <span style="color: #008000;">'x'</span>
    <span style="color: #0000FF;">elif</span> exp != snp <span style="color: #0000FF;">and</span> exp == wt:
        <span style="color: #0000FF;">return</span> <span style="color: #008000;">'X'</span>
    <span style="color: #0000FF;">else</span>:
        <span style="color: #0000FF;">return</span> <span style="color: #008000;">'-'</span>


<span style="color: #0000FF;">def</span> <span style="color: #006699;">index_seq</span><span style="color: #707183;">(</span>sequence<span style="color: #707183;">)</span>:
    <span style="color: #008000;">"""D&#233;termine la position dans la s&#233;quence exp&#233;rimentale"""</span>

    <span style="color: #BA36A5;">position</span> = <span style="color: #D0372D;">0</span>
    <span style="color: #BA36A5;">position_list</span> = <span style="color: #707183;">[]</span>
    <span style="color: #0000FF;">for</span> index, base <span style="color: #0000FF;">in</span> <span style="color: #006FE0;">enumerate</span><span style="color: #707183;">(</span>sequence<span style="color: #707183;">)</span>:
        <span style="color: #0000FF;">if</span> base != <span style="color: #008000;">'-'</span>: <span style="color: #BA36A5;">position</span> += <span style="color: #D0372D;">1</span>
        position_list.append<span style="color: #707183;">(</span><span style="color: #006FE0;">str</span><span style="color: #7388D6;">(</span>position<span style="color: #7388D6;">)</span><span style="color: #707183;">)</span>

    <span style="color: #0000FF;">return</span> position_list


<span style="color: #0000FF;">def</span> <span style="color: #006699;">polarity_snp</span><span style="color: #707183;">(</span>align<span style="color: #707183;">)</span>:
    <span style="color: #008000;">"""D&#233;termine la polarit&#233; des SNP, WT ou GS."""</span>
    <span style="color: #BA36A5;">muscle_output</span> = parse_muscle<span style="color: #707183;">(</span>align<span style="color: #707183;">)</span>

    <span style="color: #BA36A5;">consensus</span> = <span style="color: #707183;">[]</span>
    <span style="color: #0000FF;">for</span> i, base <span style="color: #0000FF;">in</span> <span style="color: #006FE0;">enumerate</span><span style="color: #707183;">(</span>muscle_output<span style="color: #7388D6;">[</span><span style="color: #008000;">'exp'</span><span style="color: #7388D6;">]</span><span style="color: #707183;">)</span>:
        <span style="color: #BA36A5;">exp</span> = muscle_output<span style="color: #707183;">[</span><span style="color: #008000;">'exp'</span><span style="color: #707183;">][</span>i<span style="color: #707183;">]</span>
        <span style="color: #BA36A5;">snp</span> = muscle_output<span style="color: #707183;">[</span><span style="color: #008000;">'snp'</span><span style="color: #707183;">][</span>i<span style="color: #707183;">]</span>
        <span style="color: #BA36A5;">wt</span> = muscle_output<span style="color: #707183;">[</span><span style="color: #008000;">'wt'</span><span style="color: #707183;">][</span>i<span style="color: #707183;">]</span>

        <span style="color: #BA36A5;">consensus</span> += are_the_same<span style="color: #707183;">(</span>
            exp, snp, wt<span style="color: #707183;">)</span>  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">+ ',' + exp + '\n' # + sortie de parse_phd</span>

    <span style="color: #0000FF;">return</span> consensus


<span style="color: #0000FF;">def</span> <span style="color: #006699;">parse_phd</span><span style="color: #707183;">(</span>align<span style="color: #707183;">)</span>:
    <span style="color: #008000;">"""Renvoit une DataFrame &#224; partir de la sortie phd de phred.</span>

<span style="color: #008000;">    """</span>
    <span style="color: #0000FF;">return</span> pd.read_table<span style="color: #707183;">(</span>align,
                         sep=<span style="color: #008000;">" "</span>,
                         skiprows=<span style="color: #D0372D;">20</span>, <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">FRAGILE. peut &#234;tre utiliser une autre fonction</span>
                         <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">pour d&#233;finir o&#249; commence la s&#233;quence, ou pour</span>
                         <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">nettoyer la sortie phd de phred</span>
                         skipfooter=<span style="color: #D0372D;">3</span>, <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">idem, moins sensible normalement.</span>
                         engine=<span style="color: #008000;">'python'</span>, <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">n&#233;cessaire pour skipfooter</span>
                         names=<span style="color: #7388D6;">[</span><span style="color: #008000;">'base'</span>, <span style="color: #008000;">'qual'</span>, <span style="color: #008000;">'phase'</span><span style="color: #7388D6;">]</span><span style="color: #707183;">)</span> <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">'phase' est la position dans</span>
    <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">le spectrogramme.</span>


<span style="color: #6434A3;">@phruscle.command</span><span style="color: #707183;">(</span><span style="color: #008000;">'table'</span>, short_help=<span style="color: #008000;">"Make a table of SNP qual and polarity"</span><span style="color: #707183;">)</span>
<span style="color: #6434A3;">@click.option</span><span style="color: #707183;">(</span><span style="color: #008000;">'-i'</span>, <span style="color: #008000;">'--input'</span>, <span style="color: #006FE0;">help</span>=<span style="color: #008000;">"Input file in abi format"</span><span style="color: #707183;">)</span>
<span style="color: #6434A3;">@click.option</span><span style="color: #707183;">(</span><span style="color: #008000;">'-o'</span>,
              <span style="color: #008000;">'--output'</span>,
              <span style="color: #006FE0;">type</span>=click.File<span style="color: #7388D6;">(</span><span style="color: #008000;">'wb'</span><span style="color: #7388D6;">)</span>,
              default=<span style="color: #008000;">'-'</span>,
              required=<span style="color: #D0372D;">False</span>,
              <span style="color: #006FE0;">help</span>=<span style="color: #008000;">"Output file in csv format"</span><span style="color: #707183;">)</span>
<span style="color: #0000FF;">def</span> <span style="color: #006699;">clean_output</span><span style="color: #707183;">(</span><span style="color: #006FE0;">input</span>, output<span style="color: #707183;">)</span>:
    <span style="color: #008000;">""""""</span>
    <span style="color: #0000FF;">assert</span> is_abi<span style="color: #707183;">(</span><span style="color: #006FE0;">input</span><span style="color: #707183;">)</span>, <span style="color: #008000;">"Input is not an abi file"</span>

    <span style="color: #BA36A5;">parse_align</span> = parse_muscle<span style="color: #707183;">(</span><span style="color: #006FE0;">input</span> + <span style="color: #008000;">".seq.aln"</span><span style="color: #707183;">)</span>

    <span style="color: #BA36A5;">clean_data</span> = pd.DataFrame<span style="color: #707183;">(</span><span style="color: #7388D6;">{</span>
        <span style="color: #008000;">'seqp'</span>: index_seq<span style="color: #909183;">(</span>parse_align<span style="color: #709870;">[</span><span style="color: #008000;">'exp'</span><span style="color: #709870;">]</span><span style="color: #909183;">)</span>,
        <span style="color: #008000;">'refp'</span>: index_seq<span style="color: #909183;">(</span>parse_align<span style="color: #709870;">[</span><span style="color: #008000;">'wt'</span><span style="color: #709870;">]</span><span style="color: #909183;">)</span>,
        <span style="color: #008000;">'cons'</span>: polarity_snp<span style="color: #909183;">(</span><span style="color: #006FE0;">input</span> + <span style="color: #008000;">".seq.aln"</span><span style="color: #909183;">)</span>
    <span style="color: #7388D6;">}</span><span style="color: #707183;">)</span>

    <span style="color: #BA36A5;">with_phd</span> = pd.concat<span style="color: #707183;">(</span>
        <span style="color: #7388D6;">[</span>
            clean_data<span style="color: #909183;">[</span>clean_data.cons != <span style="color: #008000;">'-'</span><span style="color: #909183;">]</span>.reset_index<span style="color: #909183;">(</span>
                drop=
                <span style="color: #D0372D;">True</span><span style="color: #909183;">)</span>,  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">ne garde que les positions qui existent dans le phd file</span>
            parse_phd<span style="color: #909183;">(</span><span style="color: #006FE0;">input</span> + <span style="color: #008000;">".phd.1"</span><span style="color: #909183;">)</span>
        <span style="color: #7388D6;">]</span>,
        axis=<span style="color: #D0372D;">1</span>,
        ignore_index=<span style="color: #D0372D;">True</span><span style="color: #707183;">)</span>
    <span style="color: #BA36A5;">with_phd.columns</span> = <span style="color: #707183;">[</span><span style="color: #008000;">'cons'</span>, <span style="color: #008000;">'refp'</span>, <span style="color: #008000;">'seqp'</span>, <span style="color: #008000;">'base'</span>, <span style="color: #008000;">'qual'</span>, <span style="color: #008000;">'phase'</span><span style="color: #707183;">]</span>

    <span style="color: #0000FF;">if</span> output != <span style="color: #008000;">'-'</span>:
        with_phd.to_csv<span style="color: #707183;">(</span>output, index=<span style="color: #D0372D;">False</span><span style="color: #707183;">)</span>
    <span style="color: #0000FF;">else</span>:
        click.echo<span style="color: #707183;">(</span>with_phd, <span style="color: #006FE0;">file</span>=output<span style="color: #707183;">)</span>

<span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">(X) if __name__ == '__main__':</span>
<span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">(X)     run_phruscle()</span>
</pre>
</div>

<p>
Le script suivant est un script pour déployer <code>phruscle</code>.
</p>

<div class="org-src-container">

<pre class="src src-python"><span style="color: #0000FF;">from</span> setuptools <span style="color: #0000FF;">import</span> setup, find_packages

setup<span style="color: #707183;">(</span>name=<span style="color: #008000;">"phruscle"</span>,
      version=<span style="color: #008000;">"0.0.1"</span>,
      py_modules=<span style="color: #7388D6;">[</span><span style="color: #008000;">"phruscle"</span><span style="color: #7388D6;">]</span>,
      packages=find_packages<span style="color: #7388D6;">(</span><span style="color: #008000;">'src'</span><span style="color: #7388D6;">)</span>,
      package_dir=<span style="color: #7388D6;">{</span><span style="color: #008000;">''</span>: <span style="color: #008000;">'src'</span><span style="color: #7388D6;">}</span>,
      install_requires=<span style="color: #7388D6;">[</span>
          <span style="color: #008000;">'Click'</span>,
          <span style="color: #008000;">'Biopython'</span>,
          <span style="color: #008000;">'numpy'</span>,
          <span style="color: #008000;">'pandas'</span>,
      <span style="color: #7388D6;">]</span>,
      entry_points=<span style="color: #008000;">'''</span>
<span style="color: #008000;">        [console_scripts]</span>
<span style="color: #008000;">        phruscle=phruscle:phruscle</span>
<span style="color: #008000;">    '''</span>, <span style="color: #707183;">)</span>
</pre>
</div>
<p>
Le script est là : <a href="./src/phruscle">./src/phruscle</a>.
</p>

<p>
Avec la commande suivante :
</p>
<div class="org-src-container">

<pre class="src src-sh">phruscle table -i pW85-1073.ab1 -o test.csv
</pre>
</div>

<p>
Je produit une table csv, que j'analyse rapidement dans R.
</p>

<div class="org-src-container">

<pre class="src src-R"><span style="color: #D0372D;">library</span><span style="color: #707183;">(</span>dplyr<span style="color: #707183;">)</span>
<span style="color: #D0372D;">library</span><span style="color: #707183;">(</span>ggplot2<span style="color: #707183;">)</span>

<span style="color: #006699;">setwd</span><span style="color: #707183;">(</span><span style="color: #008000;">"./tmp"</span><span style="color: #707183;">)</span>
test <span style="color: #D0372D;">&lt;-</span> <span style="color: #006699;">read.csv</span><span style="color: #707183;">(</span><span style="color: #008000;">"test.csv"</span><span style="color: #707183;">)</span> <span style="color: #D0372D;">%&gt;%</span> <span style="color: #006699;">tbl_df</span><span style="color: #707183;">()</span>

test <span style="color: #D0372D;">%&gt;%</span>
  <span style="color: #006699;">qplot</span><span style="color: #707183;">(</span>data<span style="color: #D0372D;">=</span> ., x <span style="color: #D0372D;">=</span> refp, y <span style="color: #D0372D;">=</span> qual, geom <span style="color: #D0372D;">=</span> <span style="color: #006699;">c</span><span style="color: #7388D6;">(</span><span style="color: #008000;">"point"</span>, <span style="color: #008000;">"line"</span><span style="color: #7388D6;">)</span>, color <span style="color: #D0372D;">=</span> qual<span style="color: #707183;">)</span>
</pre>
</div>

<p>
À priori tout est bon. Le but est maintenant d'utiliser ce script sur un grand
nombre de séquence.
</p>
</div>
</div>

<div id="outline-container-sec-6" class="outline-2">
<h2 id="sec-6"><span class="section-number-2">6</span> Tests</h2>
<div class="outline-text-2" id="text-6">
<p>
La fonction <i>assert</i> provient de là :
<a href="http://tldp.org/LDP/abs/html/debugging.html#ASSERT">http://tldp.org/LDP/abs/html/debugging.html#ASSERT</a>
</p>

<p>
Le script en question vient de là : <a href="https://github.com/lehmannro/assert.sh">https://github.com/lehmannro/assert.sh</a>
</p>

<div class="org-src-container">

<pre class="src src-sh">wget https://raw.github.com/lehmannro/assert.sh/v1.1/assert.sh -O test/assert.sh
chmod +x test/assert.sh
</pre>
</div>

<p>
Le script pour tester les résultats, via <code>make test</code>.
</p>

<div class="org-src-container">

<pre class="src src-sh">  <span style="color: #8D8D84;">#</span><span style="color: #8D8D84; font-style: italic;">!/bin/bash</span>

  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">cette fonction renvoit le nombre de s&#233;quence dans un fichier fastq pool&#233;.</span>
  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">elle divise le nombre de ligne par 4 pour conna&#238;tre le nombre de s&#233;quence.</span>
  check_fastq_length <span style="color: #707183;">()</span> <span style="color: #707183;">{</span>
      cat $<span style="color: #D0372D;">1</span> | <span style="color: #006FE0;">echo</span> $<span style="color: #7388D6;">(</span><span style="color: #909183;">(</span><span style="color: #FF1493;">`wc -l`</span> / <span style="color: #D0372D;">4</span><span style="color: #909183;">)</span><span style="color: #7388D6;">)</span>
  <span style="color: #707183;">}</span>

  <span style="color: #006FE0;">set</span> -e

  <span style="color: #006FE0;">.</span> ./test/assert.sh

  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">v&#233;rifie qu'il y ait bien 95 s&#233;quence dans le fastq SW et 96 dans le fastq WS</span>
  assert <span style="color: #008000;">"echo `check_fastq_length data/sw/sw_untrimmed.fastq`"</span> <span style="color: #008000;">"95"</span>
  assert <span style="color: #008000;">"echo `check_fastq_length data/ws/ws_untrimmed.fastq`"</span> <span style="color: #008000;">"96"</span>

  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">v&#233;rifie qu'yl y ait bien 95 s&#233;quences dans le fastq SW trimm&#233; et 96</span>
  <span style="color: #8D8D84;"># </span><span style="color: #8D8D84; font-style: italic;">dans le fastq WS trimm&#233;</span>
  assert <span style="color: #008000;">"echo `check_fastq_length data/sw/sw_untrimmed.fastq.trim`"</span> <span style="color: #008000;">"90"</span>
  assert <span style="color: #008000;">"echo `check_fastq_length data/ws/ws_untrimmed.fastq.trim`"</span> <span style="color: #008000;">"90"</span>
  assert <span style="color: #008000;">"echo `wc -l data/snp_call/snp_calling.dat | awk '{print $1}'`"</span> <span style="color: #008000;">"1875"</span>

  assert_end results
</pre>
</div>
</div>
</div>

<div id="outline-container-sec-7" class="outline-2">
<h2 id="sec-7"><span class="section-number-2">7</span> Dépendances</h2>
<div class="outline-text-2" id="text-7">
</div><div id="outline-container-sec-7-1" class="outline-3">
<h3 id="sec-7-1"><span class="section-number-3">7.1</span> BBmap</h3>
<div class="outline-text-3" id="text-7-1">
<p>
Le site de téléchargement est là :
<a href="https://sourceforge.net/projects/bbmap">https://sourceforge.net/projects/bbmap</a>
</p>

<p>
Pour installer bbduk :
</p>

<div class="org-src-container">

<pre class="src src-sh">  wget https://sourceforge.net/projects/bbmap/files/latest/download <span style="color: #008000;">\</span>
       -o bin/bbmap.tar.gz
  <span style="color: #006FE0;">cd</span> bin
  tar xzf bin/bbmap.tar.gz
</pre>
</div>
</div>
</div>

<div id="outline-container-sec-7-2" class="outline-3">
<h3 id="sec-7-2"><span class="section-number-3">7.2</span> R markdown</h3>
<div class="outline-text-3" id="text-7-2">
<p>
Il faut une version récente du package <code>rmarkdown</code>.
</p>

<div class="org-src-container">

<pre class="src src-R"><span style="color: #006699;">install.package</span><span style="color: #707183;">(</span><span style="color: #008000;">'rmarkdown'</span>, type <span style="color: #D0372D;">=</span> <span style="color: #008000;">"source"</span><span style="color: #707183;">)</span>
</pre>
</div>
</div>
</div>
</div>
</div><div class="col-md-3"><nav id="table-of-contents">
<div id="text-table-of-contents" class="bs-docs-sidebar">
<ul class="nav">
<li><a href="#sec-1">1. Dossiers et fichiers</a></li>
<li><a href="#sec-2">2. Contrôle qualité</a></li>
<li><a href="#sec-3">3. Analyse rapide du SNP calling de GATC</a></li>
<li><a href="#sec-4">4. SNP calling global</a>
<ul class="nav">
<li><a href="#sec-4-1">4.1. trimming des séquences</a></li>
<li><a href="#sec-4-2">4.2. Contrôle qualité des séquences trimmées</a></li>
<li><a href="#sec-4-3">4.3. Pooling des séquences dans un seul fastq</a></li>
<li><a href="#sec-4-4">4.4. Analyses des alignements via seaview</a></li>
<li><a href="#sec-4-5">4.5. snp calling via ssahaSNP</a></li>
<li><a href="#sec-4-6">4.6. création de la table d'identification des clones</a></li>
<li><a href="#sec-4-7">4.7. Analyses des résultats de snp calling</a></li>
</ul>
</li>
<li><a href="#sec-5">5. SNP calling précis <code>[0%]</code></a></li>
<li><a href="#sec-6">6. Tests</a></li>
<li><a href="#sec-7">7. Dépendances</a>
<ul class="nav">
<li><a href="#sec-7-1">7.1. BBmap</a></li>
<li><a href="#sec-7-2">7.2. R markdown</a></li>
</ul>
</li>
</ul>
</div>
</nav>
</div></div></div>
<footer id="postamble" class="">
<div><p class="author">Auteur: Samuel Barreto</p>
<p class="date">Created: 2016-04-04 Mon 18:26</p>
<p class="creator"><a href="http://www.gnu.org/software/emacs/">Emacs</a> 24.5.1 (<a href="http://orgmode.org">Org-mode</a> 8.3.4)</p>
</div>
</footer>
</body>
</html>
